Added proposal for auto-rebalance on imbalanced cluster feature in operator (#211)
strimzi/proposals: 136-auto-rebalance-on-imbalanced-clusters.md

Reviewer: * How will we track it? Is it described somewhere? Maybe I missed it.
* Why don't we expect the inter-broker rebalance to fix the inter-broker issue? My understanding is that the intra-broker rebalancing does only rebalancing between disks of the same broker. But that does not mean that inter-broker rebalancing does not solve intra-broker issues. When does CC actually decide to raise intra-broker goal violation?

Author: >When does CC actually decide to raise intra-broker goal violation?

CC only detects the intrabroker violation if we add the INTRA_BROKER_DISK_CAPACITY_GOAL` and `INTRA_BROKER_DISK_USAGE_DISTRIBUTION_GOAL` to the goal list of anomaly detector. 

> Why don't we expect the inter-broker rebalance to fix the inter-broker issue

The above goals are not taken into consideration when we are doing inter broker rebalance as far as I can think of. I can again take a look at the code if it does

Author: > How will we track it? Is it described somewhere

Handling Intra-Broker Disk Imbalances (JBOD Storage) -> This section contains all the information for the same

Reviewer: > > Why don't we expect the inter-broker rebalance to fix the inter-broker issue
> 
> The above goals are not taken into consideration when we are doing inter broker rebalance as far as I can think of. I can again take a look at the code if it does

Right, they do not do only intra-broker rebalance. My understanding always was that the intra-broker balancing does ONLY intra-broker movements. But inter-broker balancing might also do intra-broker movements.

But even if inter-broker does not do any intra-broker movements, it might resolve the intra-broker goal violation by moving partitions between brokers. Imagine that:
* Broker A has disk one that is almost full and disk two almost empty
* Broker B has disk one almost full and disk two almost empty
* -> partitions are moved from disk one of Broker A to disk two of broker B and vice versa
* -> intra-broker imbalance is solved through inter-broker rebalancing

So my understanding is pretty much that intra-broker rebalance might nor solve inter-broker anomalies, but inter-broker rebalance can easily solve intra-broker anomalies.

Reviewer 2: Will these 2 rebalance operations be done in a single reconciliation? Or will inter-broker rebalance done first and the intra-broker rebalance will be done in the next reconciliation if the goal violations still exist?

Author: For your question @tinaselenge and also @scholzj thoughts, as per the logic I stated in the proposal, I planned to do a sequential rebalance which would mean if there are both Interbroker and Intrabroker rebalance detected, we will first fix interbroker violation and then Intrabroker. Earlier I was thinking that we should not query CC again and just do the intrabroker violation as well but I think that doesn't make sense. As @scholzj pointed out, interbroker rebalance can definitely effect the cluster and resolve the intrabroker rebalance as well. So maybe the best way(in case if both anomalies are detected):

1. We first do the interbroker rebalance and assuming that the anomaly is fixed we move the state to `Idle`
2. In case, the intrabroker violation is detected again later by the, then we will fix it.

So @tinaselenge the answer to you question would be that we do not plan to do everything in a single reconciliation and @scholzj I think the above approach might resolve your query as well.

Author: @ppatierno wdyt about the above approach?

Reviewer 3: +1 from me

We will process only the most recent goal violation (the first element in the list when sorted by `detectionDate` in descending order) because earlier violations in the list may have already been resolved by previous rebalance operations.
We will parse the data from Cruise Control into a JSON object and retrieve the `detectionDate` from the most recent entry in the `recentGoalViolations` field.
This `detectionDate` timestamp will be used to determine whether a new auto-rebalance should be triggered.

#### Logic to trigger rebalance

To determine whether a detected anomaly should trigger a new rebalance, we need to compare the anomaly detection timestamp against the completion time of the most recent rebalance operation (whether manual or automatic).
This ensures we don't retrigger rebalances for goal violations that were already addressed.

**Tracking Rebalance Completion Times:**

Since auto-generated `KafkaRebalance` resources are deleted upon completion or failure, and manual `KafkaRebalance` resources can be deleted by users, we need a persistent location to track completion times.
The operator uses a **ConfigMap** to store this information.

> **ConfigMap: `<cluster-name>-auto-rebalance-imbalance-tracker`**

Reviewer 3: AFAIU, the ConfigMap is used for tracking completion time even for manual rebalancing so I guess the proposed name can't work, it should not mention auto-rebalance at all. Maybe something like `<cluster-name-rebalancing-tracker` looks better?

This ConfigMap:
- Stores the timestamp of the most recent rebalance completion
- Tracks the current rebalancing phase for JBOD clusters (inter-broker vs. intra-broker)
- Is created in the same namespace as the Kafka cluster
- Persists even after `KafkaRebalance` resources are deleted
- Allows the operator to recover state after crashes or restarts
- Is managed exclusively by the `KafkaAssemblyOperator` (maintains proper ownership boundaries)

**ConfigMap Lifecycle:**

- **Creation**: Created automatically by the operator when the first rebalance (manual or auto-generated) completes
- **Updates**: Updated throughout the lifecycle whenever rebalances complete or JBOD rebalancing phases change
- **Persistence**: The ConfigMap is **not deleted** when rebalances complete or when the Kafka cluster is idle

> - **Deletion**: Only deleted when the Kafka cluster itself is deleted (via owner reference to the Kafka CR)

Reviewer 3: or even when the `cruiseControl` field is removed from the `Kafka` custom resource and cruise control is deleted and disabled?

After the scale operation completes, if the cluster is still imbalanced, Cruise Control will detect the goal violation again in a future reconciliation, and a fresh imbalance rebalance will be triggered.
This prevents wasted work: scale operations change the cluster topology, making any in-progress imbalance rebalance based on outdated information.

#### Detecting Active Rebalances

To determine if a rebalance is already ongoing, the operator uses a unified approach that works for both manual and auto-generated `KafkaRebalance` resources.
Every reconciliation, the operator will:

1. **List all `KafkaRebalance` resources** for the cluster using `kafkaRebalanceOperator.listAsync(namespace, Labels.fromMap(Map.of(Labels.STRIMZI_CLUSTER_LABEL, clusterName)))`
2. **Extract the state** of each `KafkaRebalance` using the existing `KafkaRebalanceUtils.rebalanceState()` utility method
3. **Check for active states**: A rebalance is considered "active" if any `KafkaRebalance` resource is in one of these states:
   - `New` - Resource just created, waiting for operator processing
   - `PendingProposal` - Cruise Control is generating the rebalance proposal

>    - `ProposalReady` - Proposal generated, waiting for user approval (for manual rebalances)

Reviewer 4: So this means that if a user posts a manual rebalance, a proposal is prepared and then the user forgets about it they have effectively blocked auto-rebalancing? 

I think we should ignore `ProposalReady` and if any manual rebalances are in this state when an auto-rebalance happens they should be put into `NotReady` or whatever action is appropriate, maybe even deleted.

Reviewer: I think that is a good point. But it should likely include also ProposalPending and New, or? Or why would you exclude only ProposalReady?

Author: Wouldnt that be us basically saying that we wont support any manual rebalances during auto-rebalance?

Reviewer 3: > then the user forgets about it they have effectively blocked auto-rebalancing

Isn't it user fault?

> they should be put into NotReady or whatever action is appropriate, maybe even deleted.

I don't think we should automatically delete a resource created by the user. Not sure about changing its state either.

I am not sure I agree. Assuming the user wants to run a manual rebalance, they create the KafkaRebalance resource and wait for the proposal to be ready, then suddenly the auto-rebalance kicks in and it cancels what the user was doing on purpose. I do still think that forgetting about the KafkaRebalance being there is a user fault and we should keep considering the above state as active rebalancing. Otherwise if the user wants to run a manual rebalance they should remember to disable auto-rebalancing otherwise it could cancel the operation they want to run manually.

Reviewer: We should definitely not delete user's rebalance. I also do not think such rebalance resource has to be forgotten. It might be just sitting there and waiting for _when needed_.

But why can't we ignore it? AFAIK only one rebalance can be in process at a point in time. So if the auto-rebalance gets started first, user's rebalance will simply have to wait.

Reviewer 4: I really think we should ignore `ProposalPending/ProposalReady` manual rebalances when considering if there are in-progress rebalances and cancel any manual rebalances we find in that state. You could add an explanation condition or log to explain why that was done. I agree that we shouldn't delete the manual rebalance CR.

Auto-rebalance should take precedence, not manual. If a user leaves a manual rebalance hanging and an auto-rebalance is triggered and cancels the manual one, the chances are that the auto one will address whatever issue the manual rebalance was created to solve anyway. If it doesn't, then the user can trigger the manual one to fix whatever the auto missed.

The reverse situation (manual can block auto) would be far worse, a user creates a manual rebalance and is waiting for CC to create a proposal, gets distracted, goes for lunch etc, a disk distribution issue happens, the the auto-rebalance is blocked and the cluster goes down (or more likely alerts are firing and an incident is called that could have been avoided).

I think that if a user wants to do a specific, bespoke manual rebalance operation, then it reasonable that they should have to disable auto-rebalancing to make sure it happens uninterrupted.

Author: I have now decribed the process in a small section

- **Optional field**: If not specified, rebalancing triggers immediately when anomalies are detected

**Configuration Options:**

1. **With maintenance windows configured**:
   ```yaml
   spec:
     maintenanceTimeWindows:
       - "* * 2-4 * * ?"      # 2:00-4:59 AM UTC daily
       - "* * 14-16 * 5 ?"    # 2:00-4:59 PM UTC on Fridays
     cruiseControl:
       autoRebalance:
         - mode: imbalance
           template:
             name: my-template
   ```
   - ✅ Rebalancing and certificate renewals respect the same time windows
   - ✅ Simpler configuration - one set of windows for all maintenance activities
   - ✅ Consistent maintenance scheduling across all operator operations

2. **Without maintenance windows (immediate rebalancing)**:
   ```yaml
   spec:
     # maintenanceTimeWindows not specified
     cruiseControl:
       autoRebalance:
         - mode: imbalance
   ```
   - Rebalancing triggers immediately when anomalies are detected
   - No time-based restrictions
   - Suitable for development/test clusters or critical production clusters requiring immediate action

**Behavior:**

- **When maintenance windows are configured**:

>   - When an anomaly is detected outside a maintenance window, the operator records the detection but does NOT trigger a rebalance

Reviewer 4: Where is "the operator" recording this detection? What does it use this for? Below you say it checks CC state endpoint?

Author: By record I meant it sees the anomaly in the JSON returned by CC but doesnt take any action

Author: I will use a better term in case its confusing

Author: I have now made use of better wording

We will add labels to the metrics to differentiate the created rebalances based on mode (i.e., whether the rebalance was triggered for `full` mode for imbalanced clusters, `add-broker` mode, or `remove-broker` mode).
Cruise Control currently does not have any metrics to depict what type of anomaly was detected and when.
Therefore, we will add a new counter metric named `strimzi_auto_rebalance_anomalies_detected_total` which will be incremented whenever an anomaly is detected by the anomaly detector.
We will also add labels to differentiate anomalies based on fixability (`fixable`, `unfixable`, `mixed`) and type of anomaly (`goal_violation`, `broker_failure`, `disk_failure`, etc.).
These metrics will be exposed by the operator and will only be available when the user has configured the `imbalance` mode of auto-rebalance.

#### Detecting Persisting Anomalies with Metrics and Alerts

An anomaly may persist after a rebalance if cluster constraints prevent full optimization, new load patterns emerge immediately, or the rebalance only partially addresses violations.

> When this happens, Cruise Control keeps the same anomaly entry with the original `detectionDate`, preventing automatic retriggering (since `detectionDate < lastRebalanceCompletionTime`).

Reviewer 4: I don't think this is correct. Not only do you say the opposite to this in the failed rebalance section above, if you check the CC code you can see a new [timestamp](https://github.com/linkedin/cruise-control/blob/04081ab50f1f0f8425c2449a737f4a475236a329/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/GoalViolationDetector.java#L166) and [instance](https://github.com/linkedin/cruise-control/blob/04081ab50f1f0f8425c2449a737f4a475236a329/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/GoalViolationDetector.java#L167) is created for each anomaly, weather this is the 1st time it has been seen or the 1000th.  

Have your verified the behaviour?

Author: @tomncooper You are correct, I did fix this in my POC, sorry to not do it here. Yes everytime the anomaly persist we get a new detction time and therfore I was planning to use the metrics to tackle this issue. We will track the no of rebalances happening  for example, in an hour

Author: I have fixed this now. Feel free to open the discussion in case you think its still not fixed

These are similar to the default goals used for `KafkaRebalance` if the users don't set the rebalance goals.
**Template Goal Validation:**

If the user specifies a rebalance template, the operator validates that the goals in the template are a subset of the anomaly detection goals configured for Cruise Control.
This ensures that the resulting rebalance proposal will actually try to address the root cause of the anomaly that was detected. 
If the anomaly is due to a disk distribution goal violation but your rebalance optimization proposal does not use the DiskDistributionGoal, then it is wrong.

**When validation occurs:**
- During every `Kafka` CR reconciliation cycle, as part of the `KafkaAutoRebalanceReconciler.reconcile()` method
- The validation happens **after** Cruise Control is deployed but **before** any anomaly detection or rebalancing logic executes
- The operator first checks if the `imbalance` mode is configured in `spec.cruiseControl.autoRebalance`
- If `imbalance` mode is enabled and a template is specified, the operator validates that:

>   - All goals in the template exist in the `anomaly.detection.goals` configuration (or defaults if not configured)

Reviewer 2: Currently `anomaly.detection.goals` configuration is forbidden, so with this proposal, it will be updated to be configurable? I think that was not specifically mentioned before this line.

Author: Yes, we will be able configure the goals with this proposal. Currently if you set the anomaly detection goals they are reverted back to the default anomaly goals

Author: I have added a line for the same now

We will process only the most recent goal violation (the first element in the list when sorted by `detectionDate` in descending order) because earlier violations in the list may have already been resolved by previous rebalance operations.
We will parse the data from Cruise Control into a JSON object and retrieve the `detectionDate` from the most recent entry in the `recentGoalViolations` field.
This `detectionDate` timestamp will be used to determine whether a new auto-rebalance should be triggered.

#### Logic to trigger rebalance

To determine whether a detected anomaly should trigger a new rebalance, we need to compare the anomaly detection timestamp against the completion time of the most recent rebalance operation (whether manual or automatic).
This ensures we don't retrigger rebalances for goal violations that were already addressed.

**Tracking Rebalance Completion Times:**

Since auto-generated `KafkaRebalance` resources are deleted upon completion or failure, and manual `KafkaRebalance` resources can be deleted by users, we need a persistent location to track completion times.

> The operator uses a **ConfigMap** to store this information.

Reviewer 2: ```suggestion
The operator will use a **ConfigMap** to store this information.
```
This is not something we do today right?

Author: Yes, I will fix the sentence

In Strimzi's current implementation of Cruise Control, fixing these issues needs to be triggered manually i.e. if the cluster is imbalanced then a user might instruct Cruise Control to move the partition replicas across the brokers in order to fix the imbalance using the `KafkaRebalance` custom resource.

Users can currently enable anomaly detection and can also [set](https://strimzi.io/docs/operators/latest/full/deploying.html#setting_up_alerts_for_anomaly_detection) the notifier to one of those included with Cruise Control (`SelfHealingNotifier`, `AlertaSelfHealingNotifier`, `SlackSelfHealingNotifier` etc.).
All the `self.healing` prefixed properties are currently disabled in Strimzi's Cruise Control integration because, initially, it was not clear how self-healing would act if pods were rolled in the middle of rebalances or how Strimzi triggered manual rebalances should interact with Cruise Control triggered self-healing ones.

> ### Proposal

Reviewer: You have a wrong header level here? It nests under `Current situation` which seems wrong. This might also impact the followup headers?

Author: fixed

spec:
  goals:
    - CpuCapacityGoal
    - NetworkInboundCapacityGoal
    - DiskCapacityGoal
    - RackAwareGoal
    - MinTopicLeadersPerBrokerGoal
    - NetworkOutboundCapacityGoal
    - ReplicaCapacityGoal
  skipHardGoalCheck: true
  # ... other rebalancing related configuration
```
When the `template` is set, the operator automatically creates (or updates) a corresponding `KafkaRebalance` custom resource (based on the template) when a fixable anomaly is detected and notified by the `state` endpoint.
The operator copies over goals and rebalancing options from the referenced `template` resource to the generated rebalancing one.
If the user has not configured the anomaly detection goals in Cruise Control section of the Kafka CR then the operator will set the default goals to be used by the anomaly detector. 
The default anomaly detection goals set by the operator are `RACK_AWARENESS_GOAL`, `MIN_TOPIC_LEADERS_PER_BROKER_GOAL`, `REPLICA_CAPACITY_GOAL`, `DISK_CAPACITY_GOAL`.
These are similar to the default goals used for `KafkaRebalance` if the users don't set the rebalance goals.

> **Template Goal Validation:**

Reviewer: ```suggestion
These are similar to the default goals used for `KafkaRebalance` if the users don't set the rebalance goals.

**Template Goal Validation:**
```

3. The value is a comma-separated string of goal names (e.g., `"RackAwareGoal,ReplicaCapacityGoal,DiskCapacityGoal"`)
4. If the key is not present or empty, the operator uses the default anomaly detection goals defined in `CruiseControlConfiguration.CRUISE_CONTROL_DEFAULT_ANOMALY_DETECTION_GOALS_LIST`.
   Currently, the list contains:
   - `RACK_AWARENESS_GOAL`
   - `MIN_TOPIC_LEADERS_PER_BROKER_GOAL`
   - `REPLICA_CAPACITY_GOAL`
   - `DISK_CAPACITY_GOAL`
5. The operator parses the goals string into a list and compares it against the template goals to ensure the template goals are a subset.

**Validation behavior:**
- If validation **passes**: Auto-rebalance proceeds normally when anomalies are detected
- If validation **fails**: 
  - A `Warning` condition is added to the `Kafka` CR status
  - Auto-rebalance for the `imbalance` mode is **suspended** (not triggered)
  - Other operations (manual rebalances, scale-up/down auto-rebalances) continue normally
  - Users must correct the template configuration to resume imbalance auto-rebalancing

The `KafkaRebalance` has 4 modes: `full`, `add-broker`, `remove-broker` and `remove-disks` mode.

> The `imbalance` mode will be mapped to the `full` mode in the generated KafkaRebalance resource which means that the generated `KafkaRebalance` custom resource will have the mode set as `full` which within the Strimzi rebalancing operator means calling the Cruise Control API to run a rebalancing taking all brokers into account.

Reviewer: ```suggestion
The `imbalance` mode will use the `full` mode in the generated KafkaRebalance resource which means that the generated `KafkaRebalance` custom resource will have the mode set as `full` which within the Strimzi rebalancing operator means calling the Cruise Control API to run a rebalancing taking all brokers into account.
```

These are different modes in different part of the code. So I do not think they map. You just use the `full` rebalance.

Author: fixed

Cruise Control currently does not have any metrics to depict what type of anomaly was detected and when.
Therefore, we will add a new counter metric named `strimzi_auto_rebalance_anomalies_detected_total` which will be incremented whenever an anomaly is detected by the anomaly detector.
We will also add labels to differentiate anomalies based on fixability (`fixable`, `unfixable`, `mixed`) and type of anomaly (`goal_violation`, `broker_failure`, `disk_failure`, etc.).
These metrics will be exposed by the operator and will only be available when the user has configured the `imbalance` mode of auto-rebalance.

#### Detecting Persisting Anomalies with Metrics and Alerts

An anomaly may persist after a rebalance if cluster constraints prevent full optimization, new load patterns emerge immediately, or the rebalance only partially addresses violations.
When this happens, Cruise Control keeps detecting the same anomaly entry with a different anomaly name and different detection time.

> To avoid rebalance loops and cluster instability, the users can make use of the metrics exposed by the operator.

Reviewer: This does not help to avoid the loops. It helps to monitor the loops. That is a different, so you should update the wording.

Author: fixed now

- **Rebalance failures**: Unfixable goals detected or repeated NotReady states requiring manual intervention

When such patterns are detected, check Kafka CR status for warnings, review `KafkaRebalance` resources, query Cruise Control's state endpoint for `balancednessScore`, and verify cluster capacity matches anomaly detection goals.

**Shipped Alert Examples:**

The Strimzi project will not include pre-configured alerts for these metrics in the default shipped alerting rules.
This decision is intentional because:
- Alert thresholds are highly environment-specific (busy production clusters vs. development clusters have different normal behavior)
- Different organizations have different tolerances for rebalancing frequency
- Users should tune alerts based on their specific cluster characteristics and operational requirements

However, the Strimzi documentation will include **example alert configurations** for common scenarios:
- Example Prometheus alert for detecting frequent auto-rebalances
- Example alert for unfixable goal violations that persist beyond a threshold
- Example alert for rebalance failures

> Users can copy and adapt these examples to their monitoring systems and customize thresholds to match their operational needs.

Reviewer: This is confusing. We do not shop any _pre-configured alerts_ and _alert examples_. We ship only the examples. They are also not part of documentation per-se. They are part of examples. So it is not clear what we will or will not ship. Will we just add some documentation that people can use the alerts? Or would we actually add the examples to the alert examples?

spec:
  goals:
    - CpuCapacityGoal
    - NetworkInboundCapacityGoal
    - DiskCapacityGoal
    - RackAwareGoal
    - MinTopicLeadersPerBrokerGoal
    - NetworkOutboundCapacityGoal
    - ReplicaCapacityGoal
  skipHardGoalCheck: true
  # ... other rebalancing related configuration
```
When the `template` is set, the operator automatically creates (or updates) a corresponding `KafkaRebalance` custom resource (based on the template) when a fixable anomaly is detected and notified by the `state` endpoint.
The operator copies over goals and rebalancing options from the referenced `template` resource to the generated rebalancing one.
If the user has not configured the anomaly detection goals in Cruise Control section of the Kafka CR then the operator will set the default goals to be used by the anomaly detector. 
The default anomaly detection goals set by the operator are `RACK_AWARENESS_GOAL`, `MIN_TOPIC_LEADERS_PER_BROKER_GOAL`, `REPLICA_CAPACITY_GOAL`, `DISK_CAPACITY_GOAL`.
These are similar to the default goals used for `KafkaRebalance` if the users don't set the rebalance goals.

**Template Goal Validation:**

> If the user specifies a rebalance template, the operator validates that the goals in the template are a subset of the anomaly detection goals configured for Cruise Control.

Reviewer 4: ```suggestion
If the user specifies a rebalance template, the operator validates that the goals in the template are a superset of the anomaly detection goals configured for Cruise Control.
```
This is a key point I missed earlier, the template goals must cover all the configured anomaly goals to make sure we actually solve the detected anomaly, but the template can have more goals than those configured for the anomaly detection.

Author: Yes, I am already fixing this based on my call with Tina.

This ensures that the resulting rebalance proposal will actually try to address the root cause of the anomaly that was detected. 
If the anomaly is due to a disk distribution goal violation but your rebalance optimization proposal does not use the DiskDistributionGoal, then it is wrong.

**When validation occurs:**
- During every `Kafka` CR reconciliation cycle, as part of the `KafkaAutoRebalanceReconciler.reconcile()` method
- The validation happens **after** Cruise Control is deployed but **before** any anomaly detection or rebalancing logic executes
- The operator first checks if the `imbalance` mode is configured in `spec.cruiseControl.autoRebalance`
- If `imbalance` mode is enabled and a template is specified, the operator validates that:

>   - All goals in the template exist in the `anomaly.detection.goals` configuration (or defaults if not configured)

Reviewer 4: ```suggestion
  - All goals in the `anomaly.detection.goals` configuration (or defaults if not configured) exist in the template goals.
```

>   - The `anomaly.detection.goals` is currently forbidden, but with this proposal, the users will be able to configure it with Cruise Control.

Reviewer 4: I don't think this is actually true? `anomaly.detection.goals` or subsets of that string are not in the `FORBIDDEN_PREFIXS` list in [CruiseControlSpec](https://github.com/strimzi/strimzi-kafka-operator/blob/8dd73ff5fab5e59ebcedad2bceb0ef6dab5f0d28/api/src/main/java/io/strimzi/api/kafka/model/kafka/cruisecontrol/CruiseControlSpec.java#L46)?

Author: Its not exactly forbidden but it works weird for now. Even if you set the anoamly detction goals they will be reverted back to some default anomaly detection goals that the operator is setting

Author: So I think using `anomaly.detection.goals` is a no-op currently and therefore I say its forbidden

Reviewer 4: ```suggestion
  - The `anomaly.detection.goals` configuration is currently ignored by the operator (which uses a hard coded default set), but with this proposal, the users will be able to configure the goals and have them applied to Cruise Control.
```

The operator can retrieve the unfixable goals from the `unfixableViolatedGoals` list in the JSON response.
When unfixable goals are detected, the operator will set a warning condition in the status of the `Kafka` CR to alert users that manual intervention is required:

```yaml
status:
  autoRebalance:
    state: Idle 
    lastTransitionTime: "2025-09-22T16:04:00Z"
    modes:
      - mode: imbalance
  conditions:
    - type: Warning
      status: "True"
      reason: AutoRebalanceOnImbalanceFailure
      message: |
        The detected goal violation contains unfixable goals: [DiskDistributionGoal]. 
        These violations typically require infrastructure-level changes such as adding more brokers, racks, or resources. 
        Auto-rebalance will not be triggered until the unfixable goals are resolved.

>         Fixable goals detected: [NetworkInboundCapacityGoal]. These will be addressed once unfixable goals are resolved.

Reviewer 5: I see that Jakub's comment is resolved - so it is now shorter? Would something like `The detected goal violation contains unfixable goals: [DiskDistributionGoal]. Please check logs for more details` make sense? (Sorry I didn't see previous state of the message).

Author: Nice catch, I will shorten the warning message

For example, if the anomaly detection goals are configured as `[RackAwareGoal, ReplicaCapacityGoal, DiskCapacityGoal]` but the rebalance template specifies `[RackAwareGoal, CpuCapacityGoal]`, the validation will fail because `CpuCapacityGoal` is not in the anomaly detection goals list.

When this validation fails, the operator will:
1. Add a warning condition to the `Kafka` CR status
2. NOT trigger any auto-rebalances for the `imbalance` mode until the configuration is corrected
3. Log a warning message indicating which goals failed validation

Example warning condition:

```yaml
status:
  autoRebalance:
    state: Idle
    lastTransitionTime: "2025-09-22T16:04:00Z"
    modes:
      - mode: imbalance
  conditions:
    - type: Warning
      status: "True"
      reason: InvalidRebalanceTemplateGoals
      message: |
        The rebalance template 'my-imbalance-rebalance-template' contains goals that are not configured in anomaly.detection.goals.
        Invalid goals: [CpuCapacityGoal, NetworkInboundCapacityGoal]
        Configured anomaly detection goals: [RackAwareGoal, ReplicaCapacityGoal, DiskCapacityGoal]

>         Please update the template to only include goals that are being monitored for anomalies.

Reviewer 2: Do we get a warning if anomaly detection goal does not exist in the KafkaRebalance template? For example, anomaly detection has CpuCapacityGoal but not in the KafkaRebalance template?

The validation rule mentioned above said:
```
All goals in the `anomaly.detection.goals` configuration (or defaults if not configured) exist in the template goals.
```

Author: anomaly detction goals are only configured in the cruise control section of the Kafka CR so its not possible to set them in the template. The warning states that the goals that are added in the `anomaly.detection.goals` of the cruise control CR should definitely be present in the rebalance template

Author: I think the wordings of the last sentence `Please update the template to only include goals that are being monitored for anomalies.` sounds a bit weird. I think I will update it say something like `Please update the template to include the goals that are being monitored for anomalies in the anomaly.detection.goals configuration of the Kafka CR`

4. **Scale Operations Take Priority**
   - Start with state `RebalanceOnImbalance` and active `KafkaRebalance`
   - Trigger scale-down by reducing broker replicas
   - Verify state transitions to `RebalanceOnScaleDown` (imbalance rebalance interrupted)

5. **Unfixable Goal Violations**
   - Mock anomaly response with non-empty `unfixableViolatedGoals` list
   - Verify operator does NOT create `KafkaRebalance` resource
   - Verify Warning condition is added to Kafka CR status listing unfixable goals

6. **Rebalance Failure Recovery**
   - Start rebalance, then simulate `KafkaRebalance` transitioning to `NotReady`
   - Verify operator transitions to `Idle` and deletes failed resource
   - In next reconciliation, mock new goal violation with later timestamp
   - Verify fresh rebalance attempt is triggered

7. **Stopped Rebalance**
   - Apply `strimzi.io/rebalance=stop` annotation on generated `KafkaRebalance`
   - Simulate transition to `Stopped` state
   - Verify `<cluster-name>-rebalancing-tracker` ConfigMap's `lastRebalanceCompletionTime` is updated
   - Verify operator transitions to `Idle` and deletes resource

> ### System Test

Reviewer 5: I think that we will need more than just one test case and we should mention the scenarios here. Right now there is just the happy path. But FMPOV it would be worth checking the "not happy paths" like - there are some fixable and unfixable goals, once they are fixed, we can continue; the path which you described when there is already some running rebalance for scale up/down with some anomalies -> so if I got it correctly, we should stop it and run the full rebalance on imbalance. Maybe there are more. We should not focus just on happy path. But we need to test how the whole system will work against real clusters. 

Yes the FSM and other parts can be unit tested. But I don't think that you will find some real issues with mocks.

Author: Yes @im-konge, I agree with your suggestions but I planned with one system test for now because the tests for this feature can be time consuming and sometimes might be flaky(not sure). It takes time for cruise control to detect the anomalies as it first need to train itself. I thing a single test can take more than 10 minutes to run and that also one when the anomaly is detcted faster else it could take nearly 15 minutes. I think we should first try to see how much time it takes to run the happy path and then move forward with the other.

Reviewer 5: I know they can be flaky and it's complicated in CC land :) I was trying to fix some of the flakes myself. But I think it should be covered, it doesn't need to be run on every regression pipeline, but have some profile as we have for `recovery` and we can run it whenever we want.

Author: Okay, I will add some ST ideas to the proposal

Reviewer 5: Thanks :)

  - **Key point**: Cruise Control will keep raising the same goal violation event periodically (with updated `detectionDate` and new `anomalyId`) until the issue is fixed. The auto-rebalance will only execute once the maintenance window is satisfied and Cruise Control raises the event during that window.

- **When no maintenance windows are configured**:
  - Rebalance is triggered immediately when a fixable goal violation is detected by Cruise Control
  - The operator still relies on Cruise Control's periodic anomaly detection (every `anomaly.detection.interval.ms`)

**Emergency Override:**

If users need immediate rebalancing despite configured maintenance windows (e.g., critical disk capacity issue), they can create a manual `KafkaRebalance` resource (which bypasses maintenance windows)

#### Interaction with Scale Operations

Scale-up and scale-down auto-rebalance operations will continue to **ignore** maintenance windows (executed immediately), while imbalance auto-rebalance respects them.

> ### Handling Intra-Broker Disk Imbalances (JBOD Storage)

Reviewer 4: This section was bothering me, so I did a bit of a deep dive into the goal violation detection code. After doing that I believe the approach in this proposal will actually cause intra-broker goals to silently report no violations, regardless of the presence of disk imbalance.

The root cause is in how [`GoalViolationDetector`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/GoalViolationDetector.java#L200-L202) builds its `ClusterModel`. When the detector runs, it calls the 3-argument overload of `KafkaCruiseControl.clusterModel()`, which delegates through [`LoadMonitor`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitor.java#L474) where `populateReplicaPlacementInfo` is hardcoded to `false`. This single boolean controls the entire disk data pipeline. When it is `false`, three things don't happen: no `Disk` objects are created on any broker (the [`Broker` constructor](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java#L79-L84) sets `_diskByLogdir = Collections.emptySortedMap()`), no logdir-to-replica mapping is fetched from Kafka via `describeLogDirs()` ([`LoadMonitor`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitor.java#L521-L524) skips the call entirely), and all [`Replica`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Replica.java#L172-L175) objects are created with `disk = null`. The resulting `ClusterModel` has no disk-level information whatsoever.

When intra-broker goals run against this disk-less model, they silently become no-ops. [`IntraBrokerDiskCapacityGoal.rebalanceForBroker()`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskCapacityGoal.java#L194) calls `broker.disks()` which returns an empty collection, so the method returns immediately with no violations detected. [`IntraBrokerDiskUsageDistributionGoal.rebalanceForBroker()`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskUsageDistributionGoal.java#L260) behaves identically: it iterates `broker.disks()`, the loop body never executes, and no violations are found. The `ClusterModelStats` computed before and after optimization both show zero unbalanced disks and zero disk utilization standard deviation, so the [`AbstractGoal.optimize()`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractGoal.java#L82-L135) comparator sees no diff and the goal passes silently. No violation ever reaches the `recentGoalViolations` in the state endpoint response.

This appears to be by design, not a bug. The disk data path is intentionally gated behind the `rebalance_disk` API parameter. If you compare the anomaly detection path with the normal rebalance path: [`ProposalsRunnable`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/ProposalsRunnable.java#L88-L93) calls the 6-argument `clusterModel()` overload and passes `_isRebalanceDiskMode` as `populateReplicaPlacementInfo`. When `rebalance_disk=true`, disk data flows through the entire pipeline and intra-broker goals work correctly. But the `GoalViolationDetector` has no mechanism to request this. It uses a different `clusterModel()` overload that doesn't accept the parameter, and the `Goal` interface's [`ModelCompletenessRequirements`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/ModelCompletenessRequirements.java) has no field for requesting disk data. Cruise Control's self-healing path also [hardcodes `SELF_HEALING_IS_REBALANCE_DISK_MODE = false`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RebalanceRunnable.java#L32), which suggests that the entire anomaly-detection-to-self-healing pipeline was designed exclusively for inter-broker operations.

This appears to be a known gap in Cruise Control. [Issue #1264](https://github.com/linkedin/cruise-control/issues/1264) ("Ability to use intra.broker.goals in anomaly detection / self healing") was opened in July 2020 requesting exactly this capability. An initial [PR #1721](https://github.com/linkedin/cruise-control/pull/1721) was submitted in October 2021 but was effectively abandoned after code review feedback in 2022. A successor [PR #2288](https://github.com/linkedin/cruise-control/pull/2288) was opened in July 2025 and is still under active review. That PR introduces a new `INTRA_BROKER_GOAL_VIOLATION` anomaly type, separate configuration properties (`anomaly.detection.intra.broker.goals`, `self.healing.intra.broker.goals`), and modifications to the `GoalViolationDetector` to build a disk-aware cluster model for intra-broker goals. The fact that this required a new anomaly type, new configuration properties, and changes to the detector's model construction strongly suggests that intra-broker anomaly detection cannot be achieved simply by adding intra-broker goals to the existing `anomaly.detection.goals` configuration.

Given that this proposal has already be approved by some maintainers, I'd suggest removing intra-broker JBOD rebalancing and focusing on inter-broker goal violations only. 

Intra-broker detection could be addressed in a follow-up, ideally after [PR #2288](https://github.com/linkedin/cruise-control/pull/2288) is merged upstream, or alternatively by using an operator-side detection mechanism such as periodically calling the Cruise Control `/proposals?rebalance_disk=true` endpoint to check whether intra-broker proposals would be generated.

3. Wait for state `RebalanceOnImbalance` with active KafkaRebalance
4. Trigger scale-down by reducing broker replicas in KafkaNodePool
5. Verify imbalance KafkaRebalance is stopped (annotation `strimzi.io/rebalance=stop`)
6. Verify imbalance KafkaRebalance is deleted
7. Verify state transitions from `RebalanceOnImbalance` to `RebalanceOnScaleDown`
8. Verify scale-down KafkaRebalance resource `<cluster>-auto-rebalancing-remove-brokers` is created
9. Wait for scale-down rebalance completion
10. Verify state transitions to `Idle`

#### 4. **Rebalance Failure Handling**

Tests operator behavior when auto-triggered rebalance fails.

**Steps:**
1. Deploy Kafka cluster with auto-rebalance `imbalance` mode
2. Create cluster imbalance
3. Wait for auto-rebalance trigger and KafkaRebalance creation
4. Simulate rebalance failure by deleting Cruise Control pod while rebalancing
5. Wait for KafkaRebalance to transition to `NotReady` state
6. Verify operator logs warning about failure
7. Verify state transitions back to `Idle`
8. Verify failed KafkaRebalance resource is deleted
9. Verify ConfigMap `lastRebalanceCompletionTime` is updated with failure timestamp

> 10. Restart Cruise Control and wait for recovery

Reviewer 5: So how the Cruise Control will be restarted? When you delete the CC Pod, it will be restarted and brought back up.

Author: Yes we delete the CC pod in between it will be restartd and brought back up

11. Wait for next anomaly detection cycle
12. Verify new rebalance attempt is triggered (detectionDate > failure timestamp)

#### 5. **JBOD Intra-Broker Rebalancing** - `testAutoRebalanceOnImbalanceJBODIntraBroker()`

Tests automatic intra-broker disk rebalancing for JBOD storage.

**Steps:**
1. Deploy Kafka cluster with JBOD storage (2+ volumes per broker)
2. Configure `anomaly.detection.goals` to include `IntraBrokerDiskCapacityGoal` and `IntraBrokerDiskUsageDistributionGoal`
3. Create uneven disk usage across volumes on same broker
4. Wait for Cruise Control to detect intra-broker violation
5. Verify inter-broker rebalance is triggered first (if inter-broker violations exist)
6. Wait for inter-broker rebalance completion
7. Verify intra-broker rebalance is triggered in next detection cycle
8. Verify generated KafkaRebalance has `rebalanceDisk: true`
9. Wait for intra-broker rebalance completion
10. Verify ConfigMap is updated

> 11. Verify state returns to `Idle`

Reviewer 5: Isn't this something that for ST would require more resources, but it can be tested on IT level? I mean, in every case we are checking if the ConfigMap is updated and if we return to `Idle`, here we are checking that the KafkaRebalance will have `rebalanceDisk: true`. That's something which can be tested on UT/IT level and we don't waste resources on that. It would be beneficial to have there some added value. Or you can combine the test cases

Author: I think an ST for this situation might be good. When we work with Intrabroker rebalances we tend to use a different type of KafkaRebalance which has the `rebalanceDisk` flag enabled and also the genraed rebalance will not have any goals since intra broker rebalance dont support any goals. The issue I can think of with this ST is that how easy it would be replicate the violation.

#### 6. **Crash Recovery from RebalanceOnImbalance State** - `testCrashRecoveryDuringImbalanceRebalance()`

Tests operator crash recovery when in `RebalanceOnImbalance` state.

**Steps:**
1. Deploy Kafka cluster with auto-rebalance `imbalance` mode
2. Create imbalance and trigger auto-rebalance
3. Wait for state `RebalanceOnImbalance` and KafkaRebalance creation
4. Simulate operator crash by deleting operator pod
5. Wait for operator to restart
6. **If KafkaRebalance still exists:** Verify operator continues monitoring from current state

> 7. **If KafkaRebalance was deleted during crash:** Verify operator transitions to `Idle` and trusts Cruise Control to re-detect violations

Reviewer 5: What do you mean by that? Maybe I didn't catch it in the proposal, but how it would be deleted during crash? Mainly in the test case, you know what you are doing, so when you delete the operator Pod, nothing will be deleted, or?

I think that the test case would be useful. But I wonder if this will not create a lot of race conditions.

> 8. Verify eventual completion or re-trigger based on cluster state

Reviewer 5: So how you determine in the test that we should wait for the completion or re-trigger it? Does that mean that the test will wait for KafkaRebalance completion like 20 minutes? If so, I would put this test to some "manual tests" checklist.

Author: I think this test will not be very complex. The sentences on line 6 and 7 are basically a part of code that I have written. So the logic check after the operator restarts check if there is a KafkaRebalance resource present or not  then we verify for completion and in case the rebalance is not there than in that case we can re trigger the rebalance

# Auto-rebalance on imbalanced clusters

This proposal introduces support for automatic partition rebalancing whenever a Kafka cluster becomes imbalanced due to unevenly distributed replicas, overloaded brokers, or similar issues.
This feature is opt-in, and when enabled, allows the Strimzi Operator to automatically execute partition rebalances through `KafkaRebalance` custom resources whenever cluster imbalances are detected by Cruise Control.

## Motivation

Currently, if the cluster is imbalanced, the user would need to manually rebalance the cluster by using the `KafkaRebalance` custom resource.

> With smaller clusters, it is feasible to fix things manually. However, for larger clusters it can be very time-consuming, or just not feasible, to fix all the imbalances manually.

Reviewer: Just a nit ... I think the motivation is correct. But isn't the effort the same for the cluster regardless of the size? It is not like you run some manual calculations or reassignments. It is still one KafkaRebalance CR, or?

> With smaller clusters, it is feasible to fix things manually. However, for larger clusters it can be very time-consuming, or just not feasible, to fix all the imbalances manually.

Reviewer 3: Split sentences on two lines.

Following the above approach will provide several advantages:
* we ensure that the operator controls all rebalance and cluster remediation operations.
* using the existing `KafkaRebalance` CR system gives more visibility into what is happening and when, which (as we don't support the Cruise Control UI) enhances observability and will also aid in debugging.

### `imbalance` mode in Strimzi's auto-rebalancing feature

The [`auto-rebalancing`](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) feature in Strimzi allows the operator to run a rebalance automatically when a Kafka cluster is scaled up (by adding brokers) or scaled down (by removing brokers).

Auto-rebalancing in Strimzi currently supports two modes:
* `add-brokers` - auto-rebalancing on scale up
* `remove-brokers` - auto-rebalancing on scale down

To leverage the automated rebalance on imbalanced clusters (those with detected goal violations), we will be introducing a new mode to the auto-rebalancing feature.
The new mode will be called `imbalance`, which means that cluster imbalance was detected and rebalancing should be applied to all the brokers.

> The imbalance mode is configured by creating a new object in the `spec.cruiseControl.autoRebalance` list with its `mode` field set to `imbalance` and the corresponding rebalancing configuration is defined as a reference to a template `KafkaRebalance` custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).

Reviewer 3: ```suggestion
The imbalance mode is configured by creating a new object in the `spec.cruiseControl.autoRebalance` list with its `mode` field set to `imbalance` and the corresponding rebalancing configuration is defined as a reference to a `KafkaRebalance` template custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
```

The operator copies over goals and rebalancing options from the referenced `template` resource to the generated rebalancing one.
If the user has not configured the anomaly detection goals in Cruise Control section of the Kafka CR then the operator would set the default goals to be used by the anomaly detector. 
The default anomaly detection goals set by the operator are `RACK_AWARENESS_GOAL`, `MIN_TOPIC_LEADERS_PER_BROKER_GOAL`, `REPLICA_CAPACITY_GOAL`, `DISK_CAPACITY_GOAL`. 
These are similar to the default goals used in for KafkaRebalance if the users don't mention the rebalance goals.
If the user specifies a rebalance template, then the goals mentioned in the rebalance template will be validated against the anomaly detection goals being used by Cruise Control.
The goals in the rebalance template should be either the same or be a subset of the anomaly detection goals being used by Cruise Control.
This is to ensure rebalances are only being run with goals that Cruise Control is checking for anomalies.

> If the goals mentioned in the templates are not a subset of the configured anomaly detection goals then we will add a warning condition in the Kafka CR regarding the failed validation.

Reviewer 3: When this validation is done? I assume, based on the current logic, the Kafka rebalance assembly operator is just going to skip the `KafkaRebalance` template. While you are suggesting that such a template won't be just skipped but reconcilied to add a warning? Also, assuming a warning is there and then an anomaly is detected, what's going to do the Kafka rebalance assembly operator if the template has a warning and can't really fix the anomaly?

Author: Added more details to it now

The `KafkaRebalance` has 4 modes: `full`, `add-broker`, `remove-broker` and `remove-disks` mode.
The `imbalance` mode will be mapped to the `full` mode in the generated KafkaRebalance resource which means that the generated `KafkaRebalance` custom resource will have the mode set as `full` which within the Strimzi rebalancing operator means calling the Cruise Control API to run a rebalancing taking all brokers into account.

> The generated `KafkaRebalance` custom resource will be called `<my-cluster-name>-auto-rebalancing-imbalance-<anomalyId>`, where the `<my-cluster-name>` part comes from the `metadata.name` in the `Kafka` custom resource, `imbalance` refers to applying the rebalance to all the brokers, and the `<anomalyId>` would be retrieved from the JSON retrieved from `state` endpoint.

Reviewer: What if there are multiple anomalies at the same time? Also, how do you recover that after downtime / restarts? Given you can have only one rebalance in progress at one time, wouldn't a static name make more sense?

Author: From what I have understood form the CC code, we can have only one anomaly detected per detection cycle as per our use case since we are using goal violation only. All the violation will come under a single anomaly per detection

Reviewer 3: But if I recall correctly from my learning on this time ago, even if CC detects the same anomaly every time (i.e. the exact same goal violation) because it was not fixed, it will assign a new anomaly ID every time and it could pollute the environment with several KafkaRebalance resources representing the same anomaly. Is it really possible? How are we going to handle it?

Author: Good question, I am afraid  that we don't have a very good way of checking if its the same anomaly hiding under different name or not. I also wonder the possibility of this case. In case, an anomaly can't be fixed by rebalance, that would mean that its an unfixable anomaly, right and CC should be able to detect that

Reviewer 3: So, we know that even if the anomaly detector manager detects more anomalies, they are put into a queue.
Then the manager goes through them one by one calling the notifier. In this proposal, the notifier would be the `NoopNotifier` so it means that each anomaly is IGNORED.
At some point the operator reconciliation will kick in and it will query CC to get anomalies. It will get a JSON with a list of ignored anomalies to pick the most recent goal violation one and start a rebalance for it.
If that's the case, then I would agree with Jakub that a static name is better. We'll always pick only one anomalies from CC, it's the operator itself to drive and make such decision. Tbh what I would do would be more using a name like `<my-cluster-name>-auto-rebalancing-goal-violation` (I am not still convinced about this "imbalance") but I am open to any name suggestion, not having a strong opinion.

Author: I have fixed this now and using a static name

kind: KafkaRebalance
metadata:
  name: my-cluster-auto-rebalancing-imbalance-<anomalyID>
  finalizers:
    - strimzi.io/auto-rebalancing
spec:
  mode: full
  goals:
    - CpuCapacityGoal
    - NetworkInboundCapacityGoal
    - DiskCapacityGoal
    - RackAwareGoal
    - MinTopicLeadersPerBrokerGoal
    - NetworkOutboundCapacityGoal
    - ReplicaCapacityGoal
  skipHardGoalCheck: true
  # ... other rebalancing related configuration
```

The operator also sets a finalizer, named `strimzi.io/auto-rebalancing`, on the generated `KafkaRebalance` custom resource.
This is needed to avoid the user, or any other tooling, deleting the resource while the auto-rebalancing is still running.
The finalizer is removed when Cruise Control indicates that the partition reassignment (rebalance) process has finished allowing the generated `KafkaRebalance` custom resource to be deleted by the operator itself.
In case the rebalance finishes with an error, the error message will be propagated to the Kafka custom resource just like we do for the `remove-broker` and `add-broker` endpoint and the generated `KafkaRebalance` will be deleted.

#### State endpoint in Cruise Control

> Cruise Control provides the `state` endpoint to query the state of Kafka Cruise Control at any time by issuing a HTTP GET request.

Reviewer: ```suggestion
Cruise Control provides the `state` endpoint to query the state of Cruise Control at any time by issuing a HTTP GET request.
```

|---------------|---------|-------------------------------------------------------------------------------------------------------------------------------------|----------------------|----------|
| substates     | list    | substates for which to retrieve state from cruise-control, available substates are analyzer, monitor, executor and anomaly_detector | all substates        | yes      |
| json          | boolean | return in JSON format or not                                                                                                        | false                | yes      |
| verbose       | boolean | return detailed state information                                                                                                   | false                | yes      |
| super_verbose | boolean | return more detailed state information                                                                                              | false                | yes      |
| doAs          | string  | propagated user by the trusted proxy service                                                                                        | null                 | yes      |

> | reason        | string  | reason for the request                                                                                                              | "No reason provided" | yes      |       

Reviewer 3: I am not sure we really need this table from Cruise Control wiki being replicated here. Just mention what are you going to use in terms of query parameters and link to the official wiki page (which you are already doing below).

For this feature, we will be making use of the two substates of the `state` endpoint which are `executor` and `anomaly_detector` substate.
The `executor` state provides us with the status of rebalance i.e. no rebalance in progress, if rebalance is already running, if rebalance is in stopped state etc.
The `anomaly_detector` state provides us with the status of the anomaly detector part of Cruise Control i.e. if a goal violation was detected, the timestamp corresponding to it, the type of anomaly etc.
For more information, refer to the [state endpoint](https://github.com/linkedin/cruise-control/wiki/rest-apis#query-the-state-of-cruise-control)

### Flow diagram depicting process of auto-rebalance on imbalance

The following diagram depicts the complete auto-rebalance on imbalance process, showing how the operator queries the Cruise Control state endpoint, detects goal violations, and triggers rebalance operations:

> ![flow-diagram-auto-rebalance-on-imbalance](images/136-auto-rebalance-imbalance.png)

Reviewer 3: From the diagram I get that on each reconciliation, the operator is going to query CC state endpoint. Then if there is a rebalancing ongoing, it's going to ignore any detected anomaly. At this point I was wonder what's the point on querying CC if a rebalancing is already ongoing? We are going to ignore it, so why querying? I think the diagram should be changed in a way that the operator is going to query CC only if there is no rebalancing running.

Author: So what I meant by the flow chart was that -> On each reconciliation, the operator is going to query CC state endpoint(using executor endpoint) to know if their are any rebalances ongoing. If there are no rebalances ongoing then we detect the anomalies. I think I should add another state in the flow doagram stating that we will again query the CC endpoint with executor endpoint.

Reviewer 4: @ShubhamRwt If a rebalance is ongoing, isn't the KRAO already going to check CC for you? When does that check happen? Can you do that check once and store the state?

Author: If a rebalance is ongoing and its triggered by auto-rebalance then we dont need to check CC as we can only check the state of the KafkaAutoRebalanceReconciler. But in case the KafkaAutoRebalanceReconciler is in Idle state then we call the CC endpoint

> ![flow-diagram-auto-rebalance-on-imbalance](images/136-auto-rebalance-imbalance.png)

Reviewer: See the comment on the diagram.

#### KafkaAutoRebalanceReconciler class using state endpoint

The operator will interact with the cruise control `state` endpoint to understand when an anomaly is detected and when it should take any action.
The `KafkaAutoRebalanceReconciler` class will hold all the logic for the auto-rebalance triggered on scale up, scale down or imbalance.

> The reconciler currently detects any sort of scale down or scale up automatically every reconciliation, we will configure it to detect any imbalances in the cluster in a similar fashion. 

Reviewer: The scaling is detected from the operator. Not from Cruise Control state. So what exactly is similar fashion about it?

Author: by similar fashion I meant that the detection will happen per reconcilation as we have for scaling

Reviewer 3: But it should be clarified, I agree with Jakub the sentence is vague. It should be improved.

Author: Sure, I will fix it

But we also need to make sure that we don't detect any goal violations when a rebalance is already ongoing (manual or automatic rebalance).
An ongoing rebalance can mean that some goal violation which might be detected in that particular reconciliation will be fixed already by the rebalance and we don't need any extra rebalance.

> For the rebalances that are triggered automatically, if the `KafkaAutoRebalanceReconciler` state machine is not in `Idle` state and is currently in `RebalanceOnScaleUp`, `RebalanceOnScaleDown`or `RebalanceOnImbalance` then it means that a rebalance is happening.

Reviewer: What will be the order of things? As I assume that scale-up upltimately also mean imbalance, I guess it matters?

Author: What I mean by the above sentence is that incase the autorebalance reconciler is not in Idle state, then we will not check for any anomalies as we know that some rebalance is happening and that rebalance might fix the anomaly.

Reviewer: I'm not sure that answers the question. When I cale from 3 to 6 brokers, I assume I have two different auto-rebalance reasons:
* Imbalanace
* Scale-up

We should have clear definition which one kicks in first. I assume the scale-up should, but it is not clear.

Author: Ahh, in that case as I said in the stage of the proposal that scale up will be given more priority than imbalance. The order should first scale down operations, then scale up and then imbalance

We can ignore the goal violation detection in those cases.

> For manual rebalances, we can't use the above mechanism, therefore to do that we will make use of the `executor` state endpoint.

Reviewer 3: So it looks like we have two ways to check a rebalancing is going on. One is leveraging Strimzi resources and their state. Another one is querying Cruise Control. I would try to have just one and being consistent. I guess even when auto-rebalancing is going on, it should be possible to get it from CC endpoints, or? This way, CC would be the source of truth.
If we want to go the opposite direction instead, by using Strimzi resources as source of truth, with both auto-rebalancing and manual rebalancing, the fact that a rebalancing is running is defined by the presence of a `KafkaRebalance` resource with a pending proposal or rebalancing state.

Reviewer 4: Yes, I agree, we need to figure out the source of truth for this. We don't want to querying the CC API multiple times if we don't need too.

Author: I now use a single source of truth

```json
{
  "ExecutorState": {
    "state": "NO_TASK_IN_PROGRESS"
  },
  "version": 1
}
```
If there is some task already ongoing, then the returned JSON would be:

```json
{
  "ExecutorState": {
    "state": "LEADER_MOVEMENT_TASK_IN_PROGRESS",
    "status": {
      "finishedLeadershipMovements": 0,
      "totalLeadershipMovements": 1,
      "maximumBrokerLeadershipMovementConcurrency": 250,
      "minimumBrokerLeadershipMovementConcurrency": 250,
      "averageBrokerLeadershipMovementConcurrency": 250.0,
      "clusterLeadershipMovementConcurrency": 1000
    },
    "triggeredSelfHealingTaskId": "c2071b83-b011-4924-8fa9-8d3cb0b2ebb9",
    "triggeredTaskReason": "Self healing for GOAL_VIOLATION: {Unfixable goal violations: {}, Fixable goal violations: {DiskUsageDistributionGoal}, Exclude brokers recently (removed: true demoted: true), Provision: OVER_PROVISIONED ([RackAwareGoal] Remove at least 1 rack with brokers.)}"
  },
  "version": 1
}
```

If there is no manual or auto-rebalance running, then we can move forward to check if there were any detected goal violations or not.
To detect any goal violations we will make use of the `state` endpoint with `anomaly_detector` substate.

> The `KafkaAutoRebalanceReconciler` will send a request to the `state` endpoint with `anomaly_detector` substate every reconciliation if no rebalance is already running.

Reviewer 3: Last two sentences say the same thing twice, also instead of being verbose I would just make explicit the URL we are going to query with corresponding parameters.

The `recentGoalViolations` list will grow as new violations are detected, but we will always be interested in the latest violated goal.
This is because the `recentGoalViolations` list contains both violations that were already fixed by previous rebalances as well as violations that still need to be addressed.
We will process only the most recent goal violation (the first element in the list when sorted by `detectionDate` in descending order) because earlier violations in the list may have already been resolved by previous rebalance operations.
We will then convert the received JSON input into a JSON object and retrieve the `detectionDate` from the most recent entry in the `recentGoalViolations` field.
This `detectionDate` timestamp will be used to determine whether a new auto-rebalance should be triggered.

#### Logic to trigger rebalance

1.	We use the `lastTransitionTime` field in the `KafkaAutoRebalanceStatus` and the detection timestamp to determine whether a rebalance should be triggered or not.
2.	Every time the autoRebalance state machine transitions—e.g., from `Idle` to `RebalanceOnImbalance` or any other state—the `lastTransitionTime` is updated.

> 3.	A rebalance on imbalance will be triggered when the anomaly detection timestamp is greater than the current `lastTransitionTime`.

Reviewer 3: Putting together 2 and 3 I think we are interested only in the `lastTransitionTime` when the auto rebalance is in `Idle`, because if it's not in `Idle`, it means that a rebalancing is running and we don't query for the state to CC as you explained before?

Author: Yes

Author: I have reworked this

> 4. If the anomaly detection timestamp is greater than the `lastTransitionTime` of the `KafkaAutoRebalanceReconciler` state machine, it means that the goal violation was detected after any sort of rebalance that happened in the cluster.

Reviewer: How do the manual relabalnces fit into this?

Reviewer 4: +1 to this, your solution needs to also cover a manually triggered rebalance

#### What happens if there are no rebalances running and an anomaly is detected (The Happy Path)

1.	We check whether the anomaly detection timestamp is greater than the `lastTransitionTime` in the `status.autoRebalance` section of the Kafka CR.
2.	If the condition passes then we will trigger a rebalance and once the rebalance is complete we will move the state of the `autoRebalance` machine to `Idle`.

#### What happens if the anomaly is detected during a running rebalance

As discussed above, we first check the state of the `autoRebalance` state machine as well as the status of any manual rebalances.
If the `autoRebalance` is not in the `Idle` state and is currently in other states like `RebalanceOnImbalance`, `RebalanceOnScaleUp`, or `RebalanceOnScaleDown`, then it means a rebalance is already running. 
Additionally, if the request to the `state` endpoint with the `executor` substate shows any task in progress, then it means a rebalance is already running.

> In both scenarios, we will simply ignore the newly detected anomaly since it might be resolved by the ongoing rebalance operation.

Reviewer 3: As mentioned with a comment in the diagram, I think it's useless to get the anomalies first and then ignoring them if a rebalancing is going on. Let's invert the flow. Wdyt?

Author: As per the written logic in the proposal, we will not do the anomaly detection unless there is no ongoing rebalance. I will fix the diagram

Reviewer 4: In which case, the scenario in this section _shouldn't_ ever happen?

Author: Yes, I should fix the sentence

> If the anomaly is not resolved by the current rebalance, the anomaly detector will detect it again in a subsequent detection cycle, and a new rebalance will be triggered at that time.

Reviewer: So, the detection time will be updated even through the anomaly was detected already before?

Author: Yes, in case the same unfxied anomaly is detcted again then we should get a new timestamp

#### What happens if some rebalance fails:

> With the new `imbalance` mode, we will be introducing two new states to the FSM called `RebalanceOnImbalance` and `RebalanceOnImbalanceNotComplete`

Reviewer 3: Why do we need an additional state for failed rebalance? IIRC there is no such a specific state for failures with auto-rebalancing on scale up/down. Or am I wrong?

Author: I think it was me overthinking that in case some failure happens and it might need a human intervention and we might not need the auto rebalance to happen again and again but I think this is not really required. I will remove it

Reviewer 4: So what state will it move to if there is an error? Idle?

Reviewer 4: Will the generate KR CR move to NotReady (or whatever the fail state for that is)? How will a human operator know this happened?

Author: The generated KR Cr would move to NotReady, yes but I think the user can check the logs but I dont think there is any other thing

Author: Is the answer clear based on my previous reply?

Reviewer 3: +1 from me

The operator can retrieve the unfixable goals from the `unfixableViolatedGoals` list in the JSON response.
When unfixable goals are detected, the operator will set a warning condition in the status of the `Kafka` CR to alert users that manual intervention is required:

```yaml
status:
  autoRebalance:
    state: Idle 
    lastTransitionTime: "2025-09-22T16:04:00Z"
    modes:
      - mode: imbalance
  conditions:
    - type: Warning
      status: "True"
      reason: UnfixableViolatedGoal
      message: |
        The detected goal violation contains unfixable goals: [DiskDistributionGoal]. 
        These violations typically require infrastructure-level changes such as adding more brokers, racks, or resources. 
        Auto-rebalance will not be triggered until the unfixable goals are resolved.

>         Fixable goals detected: [NetworkInboundCapacityGoal]. These will be addressed once unfixable goals are resolved.

Reviewer: You should probably aim to make it shorter and provide full details in the logs.

>         Fixable goals detected: [NetworkInboundCapacityGoal]. These will be addressed once unfixable goals are resolved.

Reviewer 3: What happens if the user just ignore or doesn't notice the warning condition, then on the next reconciliation, the unfixable violation is not there anymore but there is only a fixable one? Does the operator run the rebalancing by clearing this warning? Or this warning condition will block everything else?

Author: Since with every reconcilation we check for new anomaly and not the old ones. We will remove the warning

```

The warning message will include both the unfixable goals (requiring manual intervention) and any fixable goals (which will be addressed once the infrastructure constraints are resolved).

#### What happens if the rebalance template contains invalid goals

> When a user specifies a rebalance template with the `template` field, the operator validates that the goals in the template are a subset of the goals configured for anomaly detection in Cruise Control.

Reviewer 3: As mentioned in a previous comment, it's not clear when this validation happens.

Author: I have added information about it now

      message: |
        The rebalance template 'my-imbalance-rebalance-template' contains goals that are not configured in anomaly.detection.goals.
        Invalid goals: [CpuCapacityGoal, NetworkInboundCapacityGoal]
        Configured anomaly detection goals: [RackAwareGoal, ReplicaCapacityGoal, DiskCapacityGoal]
        Please update the template to only include goals that are being monitored for anomalies.
```

Users must either:
- Update the rebalance template to only include goals from the anomaly detection configuration, or
- Update the Cruise Control `anomaly.detection.goals` configuration to include the additional goals

#### Stopping a running rebalance

To stop a running rebalance, users can apply the `strimzi.io/rebalance=stop` annotation on the generated `KafkaRebalance` resource.
This will stop the running rebalance and the stopped `KafkaRebalance` resource will be deleted.
Users will need to ensure that they disable the `imbalance` mode of auto-rebalance after stopping the rebalance; otherwise, the rebalance will be triggered again since the anomalies are still present in the cluster.

#### Metrics for tracking the rebalance requests

> If users want to track when auto-rebalances happen, they can access the Strimzi [metrics](https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/grafana-dashboards/strimzi-operators.json#L712) for the `KafkaRebalance` custom resources. This includes tracking when they were visible/created.

Reviewer: Isn't that just a number? That is not much in terms of tracking.

Therefore, we will add a new counter metric named `strimzi_auto_rebalance_anomalies_detected_total` which will be incremented whenever an anomaly is detected by the anomaly detector. We will also add labels to differentiate anomalies based on fixability (`fixable`, `unfixable`, `mixed`) and type of anomaly (`goal_violation`, `broker_failure`, `disk_failure`, etc.).
These metrics will be exposed by the operator and will only be available when the user has configured the `imbalance` mode of auto-rebalance.

### Auto-rebalancing execution for `imbalance` mode

#### Auto-rebalancing Finite State Machine (FSM) for `imbalance` mode

Currently, the auto-rebalancing mechanism runs through a Finite State Machine (FSM) made by the following states:

* **Idle**: Initial state with a new auto-rebalancing initiated when scaling down/up operations were requested. This is also the ending state after an auto-rebalancing completed successfully or failed.
* **RebalanceOnScaleDown**: a rebalancing related to a scale down operation is running.
* **RebalanceOnScaleUp**: a rebalancing related to a scale up operation is running.

> As we discussed above, with the new `imbalance` mode, we will be introducing two new states to the FSM called `RebalanceOnImbalance` and `RebalanceOnImbalanceNotComplete`

Reviewer: Why is `RebalanceOnImbalanceNotComplete` needed?

Author: I think it was me overthinking that in case some failure happens and it might need a human intervention and we might not need the auto rebalance to happen again and again but I think this is not really required. I will remove it

These state will be associated with rebalances triggered by imbalanced cluster.

> With the new `imbalance` mode, the FSM state transitions would look something like this:

Reviewer: Would it look like this? Or only something like this but different?

Author: It will look like the lines I wrote below -> I will fix this sentence

> ![auto-rebalance state machine flow diagram](images/136-finite-state-machine.png)

Reviewer 3: both in the diagram and in the descriptions below there is no mention to the `RebalanceOnImbalanceNotComplete` state.

Author: I will remove this state as it is not required

  * **Idle**: if a scale down operation was requested, it was executed and completed successfully/failed or a full rebalance was asked due to an anomaly but since the scale-down rebalance is done, we can ignore the anomalies assuming they are fixed by the rebalance. In case, they are not fixed, Cruise Control will detect them again and a new rebalance would be requested.

* from **RebalanceOnScaleUp**:
  * **RebalanceOnScaleUp**: if a rebalancing on scale up is still running or another one was requested while the first one ended.
  * **RebalanceOnScaleDown**: if a scale down operation was requested, so the current rebalancing scale up is stopped (and queued) and a new rebalancing scale down is started. The rebalancing scale up will be postponed.
  * **Idle**: if a scale up operation was requested, it was executed and completed successfully/failed or a full rebalance was asked due to an anomaly but since the scale-up rebalance is done, we can ignore the anomalies assuming they are fixed by the rebalance. In case, they are not fixed, Cruise Control will detect them again and a new rebalance would be requested.

* from **RebalanceOnImbalance**:

>   * **RebalanceOnImbalance**: if another goal violation was detected while the first one ended or rebalance is still running.

Reviewer: What does this mean? That you stop the rebalance and start a new one? I do not think that was clear from the proposal itself.

Author: No it only means that in case another anomaly is detected after a previous rebalance rebalance is finished in those cases we can transition from RebalanceOnImbalance back to  RebalanceOnImbalance again. I will omit the `rebalance is still running` part since that doesnt make sense. We will ignore any anomaly in case rebalance is running

Reviewer 3: > No it only means that in case another anomaly is detected after a previous rebalance rebalance is finished in those cases we can transition from RebalanceOnImbalance back to RebalanceOnImbalance again.

But if a previous rebalance is finished we should be in Idle, not in RebalanceOnImbalance. So it's not clear to me as well.

Author: Yes @ppatierno we do hit `Idle` first, I think this should be fixed.

  * **RebalanceOnScaleUp**: if a rebalancing on scale up is in queue, then the rebalance on imbalance will be stopped and the scale up will happen first. If there is a rebalancing scale down in queue too, then it will be executed before both scale up and rebalance on imbalance.
  * **RebalanceOnScaleDown**: if a scale down operation was requested, then the rebalance on imbalance will be stopped and the scale down will be allowed to finish first and after that rebalance on imbalance will be executed.
  * **Idle**: if full rebalance was requested, it was executed and completed successfully or failed.

On each reconciliation, the following process will be used:

1. The `KafkaClusterCreator` creates the `KafkaCluster` instance.
2. The `KafkaAutoRebalancingReconciler.reconcile()` will then check for ongoing rebalances and the state of the `auto-rebalance`.

> 3. If the state is `Idle` and no rebalance is running, then we request the `state` endpoint with `anomaly_detector` substate to get the detected violations and its respective timestamp. 

Reviewer 3: This seems to be correct. First check rebalancing ongoing, then ask to CC but it's the opposite to what you said before where I left comments. Also it doesn't mention the check for manual rebalancing.

Author: I did mention about manual rebalancing that we will do it by querying the state endpoint  with executor substate. Do you want that to be mentioned here as well?

Reviewer 3: I am questioning, as in other comments, the two different approaches to check ongoing rebalancing. On one side checking the state in custom resources on the other side checking the state endpoint from CC. What I was wondering if we can consolidate with a single way, or it's not possible, or it's not the right solution.

Author: Thats a good question, can there be a case where even when we are not rebalancing, we still stay on the states like RebalanceOnScaleUp/Down. I remember something about a fix you made once, where if the scale up fails we still remain on the same state instead of moving to Idle and a new rebalance is started later. Correct me If I am wrong

Author: @ppatierno Hi, any thoughts on this?

Author: Also I think if consider an optimized approach, should we do somthing like frist checking if the state is `Idle` or not. In case it is not then we dont even call the state endpoint with the `executor` substate. That way, we can call the state endpoint only when manual rebalances happened? It also gives us an optimized apprach since in cases of autorebalance we can do fast in-memory state check instead of getting response from the endpoint?

Author: @tomncooper @scholzj what do you think about the above suggestions. Should we use both the approaches i.e. first checking the state of the reconciler and then if its idle checking for a manual rebalance by calling CC endpoint or just using one out of them which will probabaly be calling CC everytime

Reviewer: I think it needs to combine both approaches. But they need to be in some order. E.g., if the auto-rebalancing is not Idle, there is probably no need to check the endpoint. But if it is Idle, it seems to me like a good way to check if some manual rebalance is happening or not.

Author: @scholzj Yes, that is exactly the approach Iam using as depicted in the new flow diagram. Thanks for your feedback

Reviewer: Just to be clear ... whilethis approach seems to make sense to me, @ppatierno would be the SME here. So if he has better ideas, I'm happy to follow 😉

Reviewer 3: I agree on checking the status of the auto-rebalancing but what I was trying to say is that why for manual rebalancing we are looking at CC API and not if there is any `KafkaRebalance` in place (which would be anyway the same for an auto-rebalancing). This is what I was questioning. Even if there is an auto-rebalancing we can check CC instead of the state of the Kube resource. So it's valid in both ways. I was wondering to use only one way: checking Kube resources or checking CC API endpoint. But I am not against to use an hybrid approach.

Reviewer: Kyle opened the Cruise Control API in the past to Strimzi users. I guess you can say that they use it _at their own risk_. But are we sure the KafkaRebalance is the right source of truth?

Author: I have reworked the proposal to use a single approach

* if `Stopped`
  * the rebalancing is stopped, transition to **Idle** and also removing the corresponding mode from the status. The operator deletes the generated `KafkaRebalance` custom resource. The users should then disable the `imbalance` mode of auto-rebalance else a new rebalance would be triggered.

If, during an ongoing auto-rebalancing, the `KafkaRebalance` custom resource is not there anymore on the next reconciliation, it could mean the user deleted it while the operator was stopped/crashed/not running.
In this case, the FSM will assume it as falling in the `NotReady` case above.

## Affected/not affected projects

This change will affect the Strimzi cluster operator.

### Required API Changes

The following enum values need to be added to support this feature:

1. **KafkaAutoRebalanceMode enum** (`io.strimzi.api.kafka.model.kafka.cruisecontrol.KafkaAutoRebalanceMode`):
   - Add `IMBALANCE("imbalance")` to support the new imbalance mode

2. **KafkaAutoRebalanceState enum** (`io.strimzi.api.kafka.model.kafka.cruisecontrol.KafkaAutoRebalanceState`):
   - Add `RebalanceOnImbalance` - State when a rebalancing related to cluster imbalance is running

>    - Add `RebalanceOnImbalanceNotComplete` - State when a rebalancing on imbalance has failed, allowing users to understand the failure without retriggering

Reviewer: If it is API changes, maybe you should talk in terms of API paths and not Java classes.

Author: Ok I will fix this

* `remove-brokers` - auto-rebalancing on scale down

To leverage the automated rebalance on imbalanced clusters (those with detected goal violations), we will be introducing a new mode to the auto-rebalancing feature.
The new mode will be called `imbalance`, which means that cluster imbalance was detected and rebalancing should be applied to all the brokers.
The imbalance mode is configured by creating a new object in the `spec.cruiseControl.autoRebalance` list with its `mode` field set to `imbalance` and the corresponding rebalancing configuration is defined as a reference to a template `KafkaRebalance` custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
The `template` field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration (i.e. the same used for unmodified manual `KafkaRebalance` invocations).
To provide users more flexibility, they only have to configure the auto-rebalance modes they wish to use, whether it be `add-brokers`, `remove-brokers`, or `imbalance`.

> When the auto-rebalance configuration is set with `imbalance` mode enabled, the operator will trigger a partition rebalance whenever a goal violation is detected by the anomaly detector.

Reviewer 4: ```suggestion
When the auto-rebalance configuration is set with `imbalance` mode enabled, the operator will trigger a partition rebalance whenever a fixable goal violation is detected by the anomaly detector.
```

The configuration for Cruise Control rebalance is provided through a `KafkaRebalance` custom resource which is specifically defined as a template and is referenced in the `template` field in the example above.
A `KafkaRebalance` custom resource with the `strimzi.io/rebalance-template: true` annotation set can be used as a template.
When a `KafkaRebalance` with this annotation is created, the cluster operator doesn't run any rebalancing.
This is not an actual rebalance request to get an optimization proposal; it is simply where the configuration for auto-rebalancing is defined.

Here is an example template:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-imbalance-rebalance-template
  annotations:
    strimzi.io/rebalance-template: "true" # specifies that this KafkaRebalance is a rebalance configuration template
spec:
  goals:
    - CpuCapacityGoal
    - NetworkInboundCapacityGoal
    - DiskCapacityGoal
    - RackAwareGoal
    - MinTopicLeadersPerBrokerGoal
    - NetworkOutboundCapacityGoal
    - ReplicaCapacityGoal
  skipHardGoalCheck: true
  # ... other rebalancing related configuration
```

> When the `template` is set, the operator automatically creates (or updates) a corresponding `KafkaRebalance` custom resource (based on the template) when an anomaly is detected and notified by the `state` endpoint.

Reviewer 4: ```suggestion
When the `template` is set, the operator automatically creates (or updates) a corresponding `KafkaRebalance` custom resource (based on the template) when a fixable anomaly is detected and notified by the `state` endpoint.
```

The operator copies over goals and rebalancing options from the referenced `template` resource to the generated rebalancing one.
If the user has not configured the anomaly detection goals in Cruise Control section of the Kafka CR then the operator would set the default goals to be used by the anomaly detector. 
The default anomaly detection goals set by the operator are `RACK_AWARENESS_GOAL`, `MIN_TOPIC_LEADERS_PER_BROKER_GOAL`, `REPLICA_CAPACITY_GOAL`, `DISK_CAPACITY_GOAL`. 
These are similar to the default goals used in for KafkaRebalance if the users don't mention the rebalance goals.
**Template Goal Validation:**

If the user specifies a rebalance template, the operator validates that the goals in the template are a subset of the anomaly detection goals configured for Cruise Control.

> This ensures rebalances are only triggered for goals that Cruise Control actively monitors for violations.

Reviewer 4: Well, the more important reason is that the resulting rebalance proposal will actually try to address the root cause of the anomaly that was detected. If the anomaly is due to a disk distribution goal violation but your rebalance optimization proposal does not use the DiskDistributionGoal, you are gonna be in trouble.

Author: I have added this now

The following diagram depicts the complete auto-rebalance on imbalance process, showing how the operator queries the Cruise Control state endpoint, detects goal violations, and triggers rebalance operations:

![flow-diagram-auto-rebalance-on-imbalance](images/136-auto-rebalance-imbalance.png)

#### Reconciliation logic using the state endpoint

The operator will interact with the cruise control `state` endpoint to understand when an anomaly is detected and when it should take any action.
The `KafkaAutoRebalanceReconciler` class will hold all the logic for the auto-rebalance triggered on scale up, scale down or imbalance.
The reconciler currently detects any sort of scale down or scale up automatically every reconciliation.
We will configure the KafkaAutoRebalanceReconciler to detect the imbalanced cluster every reconcilation too. 
But we also need to make sure that we don't detect any goal violations when a rebalance is already ongoing (manual or automatic rebalance).
An ongoing rebalance can mean that some goal violation which might be detected in that particular reconciliation will be fixed already by the rebalance and we don't need any extra rebalance.

> For the rebalances that are triggered automatically, if the auto-rebalance state machine is not in `Idle` state and is currently in `RebalanceOnScaleUp`, `RebalanceOnScaleDown` or `RebalanceOnImbalance` then it means that a rebalance is happening.

Reviewer 4: So, following my earlier question, what updates this state? Is it an earlier part of the KRAO?

> Talking about the priority of rebalance operations, scale down rebalance operations are the most priority followed by scale down and then rebalance on imbalance.

Reviewer 4: ```suggestion
Talking about the priority of rebalance operations, scale down rebalance operations are the highest priority followed by scale down and then rebalance on imbalance.
```

```yaml
status:
  autoRebalance:
    state: Idle 
    lastTransitionTime: "2025-09-22T16:04:00Z"
    modes:
      - mode: imbalance
  conditions:
    - type: Warning
      status: "True"
      reason: UnfixableViolatedGoal
      message: |
        The detected goal violation contains unfixable goals: [DiskDistributionGoal]. 
        These violations typically require infrastructure-level changes such as adding more brokers, racks, or resources. 
        Auto-rebalance will not be triggered until the unfixable goals are resolved.
        Fixable goals detected: [NetworkInboundCapacityGoal]. These will be addressed once unfixable goals are resolved.
```

The warning message will include both the unfixable goals (requiring manual intervention) and any fixable goals (which will be addressed once the infrastructure constraints are resolved).

#### What happens if the rebalance template contains invalid goals

When a user specifies a rebalance template with the `template` field, the operator validates that the goals in the template are a subset of the goals configured for anomaly detection in Cruise Control.

> **Edge Case - Template Goals Not in Anomaly Detection Goals:**

Reviewer 4: Again, this is kind of a repeat of the paragraph above but with more details?

Author: fixed now

If, during an ongoing auto-rebalancing, the `KafkaRebalance` custom resource is not there anymore on the next reconciliation, it could mean the user deleted it while the operator was stopped/crashed/not running.
In this case, the FSM will assume it as falling in the `NotReady` case above.

### Deletion of Auto-Generated KafkaRebalance Resources

The auto-generated `KafkaRebalance` resource (`<cluster-name>-auto-rebalancing-goal-violation`) is deleted by the operator in the following scenarios:

1. **Successful completion** (`Ready` state): After the rebalance completes successfully, the operator removes the `strimzi.io/auto-rebalancing` finalizer and deletes the resource.

2. **Failure** (`NotReady` state): When the rebalance fails, the error is propagated to the `Kafka` CR status, then the operator removes the finalizer and deletes the resource. Future reconciliations may detect the same anomaly and trigger a new rebalance attempt.

> 3. **Manual stop** (`Stopped` state): If the user stops the rebalance via the `strimzi.io/rebalance=stop` annotation, the operator deletes the resource after transitioning to `Idle`. Users must disable the `imbalance` mode to prevent the rebalance from being immediately re-triggered.

Reviewer 4: Immediately? Or on the next reconciliation/anomaly detection cycle?

Author: next reconciliation cycle. I have fixed this now

# Auto-rebalance on imbalanced clusters

This proposal introduces support for automatic partition rebalancing whenever a Kafka cluster becomes imbalanced due to unevenly distributed replicas, overloaded brokers, or similar issues.

This feature is opt-in.
When enabled, it allows the Strimzi Operator to automatically execute partition rebalances through `KafkaRebalance` custom resources whenever cluster imbalances are detected by Cruise Control.

## Motivation

Currently, if the cluster is imbalanced, the user would need to manually rebalance the cluster by using the `KafkaRebalance` custom resource.

> Every cluster requires manual intervention to trigger rebalancing when imbalances are detected, regardless of its size.

Reviewer: This sentence sounds strange. I would maybe remove it. There are many people who live without rebalancing all their life. Especially if you have small cluster with 3 brokers and use replication factor 3, there is actually not much to rebalance.

Reviewer 3: I guess that Shubham meant to say something like "if your cluster is unbalanced and you want to do some rebalancing, the only way you have today is the manual one, there is no automation which is the goal for this proposal". I don't think he meant to say that all clusters of any size needs rebalancing. Maybe he can clarify the meaning.

Reviewer: Yes, I assume so, but we should clarify it.

Author: I meant that with respect to Strimzi, yes, If we want to do any rebalance due to imbalance then we have to do it manually and there is now way for auto rebalance on imbalance mechanism.

  - The template reference points to a valid `KafkaRebalance` resource with the `strimzi.io/rebalance-template: true` annotation
- This validation runs on **every reconciliation**, ensuring configuration changes are caught immediately
- Configuration errors are detected early, preventing invalid auto-rebalance attempts and providing immediate feedback to users

**How the operator retrieves anomaly detection goals:**

The `KafkaAutoRebalanceReconciler` retrieves the anomaly detection goals from the Kafka CR's Cruise Control configuration:
1. The operator accesses `kafkaCr.getSpec().getCruiseControl().getConfig()` which returns a `Map<String, Object>`
2. It looks up the `anomaly.detection.goals` key using `CruiseControlConfigurationParameters.ANOMALY_DETECTION_CONFIG_KEY.getValue()`
3. The value is a comma-separated string of goal names (e.g., `"RackAwareGoal,ReplicaCapacityGoal,DiskCapacityGoal"`)
4. If the key is not present or empty, the operator uses the default anomaly detection goals defined in `CruiseControlConfiguration.CRUISE_CONTROL_DEFAULT_ANOMALY_DETECTION_GOALS_LIST`:
   - `RACK_AWARENESS_GOAL`
   - `MIN_TOPIC_LEADERS_PER_BROKER_GOAL`
   - `REPLICA_CAPACITY_GOAL`

>    - `DISK_CAPACITY_GOAL`

Reviewer: So, are these goals what is in `CRUISE_CONTROL_DEFAULT_ANOMALY_DETECTION_GOALS_LIST`? It is a bit confusing what the list of goals means here.

Reviewer 3: Yes, Shubham is listing what is in the `CRUISE_CONTROL_DEFAULT_ANOMALY_DETECTION_GOALS_LIST`. Would you suggest to remove the list and referencing to the constant definition only (within the code) or something else? Just trying to understand what "confusing" means here.

Reviewer: I would probably just reference the list. Or one can do something like to make it more clear...

```
4. If the key is not present or empty, the operator uses the default anomaly detection goals defined in `CruiseControlConfiguration.CRUISE_CONTROL_DEFAULT_ANOMALY_DETECTION_GOALS_LIST`.
   Currently, the list contains:
     - ...
```

Author: I will try to make it simpler if it sounds confusing

5. The operator parses the goals string into a list and compares it against the template goals to ensure the template goals are a subset.

**Validation behavior:**
- If validation **passes**: Auto-rebalance proceeds normally when anomalies are detected
- If validation **fails**: 
  - A `Warning` condition is added to the `Kafka` CR status
  - Auto-rebalance for the `imbalance` mode is **suspended** (not triggered)
  - Other operations (manual rebalances, scale-up/down auto-rebalances) continue normally
  - Users must correct the template configuration to resume imbalance auto-rebalancing

The `KafkaRebalance` has 4 modes: `full`, `add-broker`, `remove-broker` and `remove-disks` mode.

> The `imbalance` mode will be mapped to the `full` mode in the generated KafkaRebalance resource which means that the generated `KafkaRebalance` custom resource will have the mode set as `full` which within the Strimzi rebalancing operator means calling the Cruise Control API to run a rebalancing taking all brokers into account.

Reviewer 3: ```suggestion
The `imbalance` mode will be mapped to the `full` mode in the generated `KafkaRebalance` resource which means that the generated `KafkaRebalance` custom resource will have the mode set as `full` which within the Strimzi rebalancing operator means calling the Cruise Control API to run a rebalancing taking all brokers into account.
```

> The generated `KafkaRebalance` custom resource will be called `<my-cluster-name>-auto-rebalancing-goal-violation`, where the `<my-cluster-name>` part comes from the `metadata.name` in the `Kafka` custom resource, `goal-violation` means that a goal related imbalance was detected in the cluster.

Reviewer: Should the mode name be in the name? E.g. `<my-cluster-name>-auto-rebalancing-imbalance`? That will make it easier to understand for the user.

Author: I named it `<my-cluster-name>-auto-rebalancing-goal-violation` since we only support goal violations for now. I think if we think about future like have fix for  topic anomalies etc then it would be great to have distinct names based on the anomaly type.

Reviewer: Isn't `imbalance` the distinct type you can name it after?

Author: It sound good too but I wanted to name it as per the anomaly. (since we are handling goal violations so `<my-cluster-name>-auto-rebalancing-goal-violation` but I am open to change the names.

Author: @ppatierno @tomncooper any preference from your side?

Author: I am using <my-cluster-name>-auto-rebalancing-imbalance for now but we can change it in case others have some other preference

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-cluster-auto-rebalancing-goal-violation
  finalizers:
    - strimzi.io/auto-rebalancing
spec:
  mode: full
  goals:
    - CpuCapacityGoal
    - NetworkInboundCapacityGoal
    - DiskCapacityGoal
    - RackAwareGoal
    - MinTopicLeadersPerBrokerGoal
    - NetworkOutboundCapacityGoal
    - ReplicaCapacityGoal
  skipHardGoalCheck: true
  # ... other rebalancing related configuration
```

> The operator also sets a finalizer, named `strimzi.io/auto-rebalancing`, on the generated `KafkaRebalance` custom resource.

Reviewer: Do we use this also for adding and removing brokers? Worth mentioning here whether it is the same pattern or not.

Reviewer 3: Yeah that's the same.

Author: Yes, thats the same. I will add a sentence to it

The operator will use this endpoint with specific substates to coordinate auto-rebalancing operations.

**Anomaly Detector Substate (`state?substates=anomaly_detector`)**

The anomaly detector substate provides information about detected anomalies in the cluster, including goal violations.
The operator queries this substate to identify when cluster imbalances occur and determine whether an auto-rebalance should be triggered.
The response includes details about fixable and unfixable goal violations, anomaly timestamps, and the current status of detected anomalies.

For more information, refer to the [state endpoint documentation](https://github.com/linkedin/cruise-control/wiki/rest-apis#query-the-state-of-cruise-control).

### Flow diagram depicting process of auto-rebalance on imbalance

The following diagram depicts the complete auto-rebalance on imbalance process, showing how the operator queries the Cruise Control state endpoint, detects goal violations, and triggers rebalance operations:

![flow-diagram-auto-rebalance-on-imbalance](images/136-auto-rebalance-imbalance.png)

#### Reconciliation logic using the state endpoint

> The operator will interact with the cruise control `state` endpoint to understand when an anomaly is detected and when it should take any action.

Reviewer: Let's use the right name everywhere please.

```suggestion
The operator will interact with the Cruise Control `state` endpoint to understand when an anomaly is detected and when it should take any action.
```

The reconciler currently detects any sort of scale down or scale up automatically every reconciliation.
We will configure the KafkaAutoRebalanceReconciler to detect the imbalanced cluster every reconciliation too. 
But we also need to make sure that we don't detect any goal violations when a rebalance is already ongoing (manual or automatic rebalance).
An ongoing rebalance can mean that some goal violation which might be detected in that particular reconciliation will be fixed already by the rebalance and we don't need any extra rebalance.

Talking about the priority of rebalance operations, scale down rebalance operations have the highest priority, followed by scale up, and then rebalance on imbalance.
The priority ordering is designed to prevent wasted work and optimize resource usage:
- **Scale-down has highest priority**: Moving partition replicas to brokers that will soon be removed wastes resources and time, so scale-down operations must complete first
- **Scale-up has second priority**: Quickly utilizing new capacity is important for cluster stability

> - **Imbalance rebalancing has lowest priority**: It addresses steady-state optimization rather than cluster topology changes, so it can wait for scaling operations to complete

Reviewer: This makes it sound like this is a queue which we will go through. Is it the case? Or is it that we will just take the highest priority task and after they are done we do the imbalance only if Cruise Control still reports it?

Reviewer 3: No there is no such a queue, I think this is trying to summarize what's going to happen in the corresponding FSM (i.e. a anomaly is skipped if there is a rebalance on scale-down, ...). I would agree it could sound like a queue handling with priorities. @ShubhamRwt because we have the FSM explained later on, I guess you could even remove this list? Wdyt @scholzj ?

Author: I added it since in one the of the previous suggestions, Jakub suggested that it would be good if we can mention the priorities i.e what takes precedence scale up or scale down  or imbalance etc. Therefore I tried to add something like this. Ic an surely remove it if it sound redundant /confusing

Reviewer: I think some rewording might help as well. For example `priority ordering` suggests to me that there is an order and that suggests the lower order items matter. But if it is not the case, it should be more clear that we pick the hifhest prio item and drop the rest and CC either raises them again or not.

2. **Extract the state** of each KafkaRebalance using the existing `KafkaRebalanceUtils.rebalanceState()` utility method
3. **Check for active states**: A rebalance is considered "active" if any KafkaRebalance resource is in one of these states:
   - `New` - Resource just created, waiting for operator processing
   - `PendingProposal` - Cruise Control is generating the rebalance proposal
   - `ProposalReady` - Proposal generated, waiting for user approval (for manual rebalances)
   - `Rebalancing` - Rebalance is actively executing partition movements

If any KafkaRebalance resource is in an active state, the operator will skip anomaly detection for imbalance mode and wait for the next reconciliation cycle.
This approach works uniformly for:
- **Auto-generated rebalances**: Created by the operator for scale-up, scale-down, or imbalance scenarios
- **Manual rebalances**: Created by users with arbitrary names

The implementation leverages the existing `KafkaRebalanceUtils` class which is already used throughout the Strimzi codebase for state extraction and validation.

> All KafkaRebalance resources (manual and auto-generated) have the `strimzi.io/cluster` label, allowing the operator to discover them via label selector.

Reviewer 3: ```suggestion
All `KafkaRebalance` resources (manual and auto-generated) have the `strimzi.io/cluster` label, allowing the operator to discover them via label selector.
```

    "metrics": {
      "meanTimeBetweenAnomalies": {
        "GOAL_VIOLATION": 3.28,
        "BROKER_FAILURE": 0.0,
        "METRIC_ANOMALY": 0.0,
        "DISK_FAILURE": 0.0,
        "TOPIC_ANOMALY": 0.0
      },
      "meanTimeToStartFix": 2890.0,
      "numSelfHealingStarted": 1,
      "numSelfHealingFailedToStart": 0,
      "ongoingAnomalyDuration": 0.0
    },
    "ongoingSelfHealingAnomaly": "c2071b83-b011-4924-8fa9-8d3cb0b2ebb9",
    "balancednessScore": 88.45
  },
  "version": 1
}
```
The `recentGoalViolations` field in the above JSON depicts which goal violations happened in the cluster and when they occurred.
The `recentGoalViolations` list will grow as new violations are detected, but we will always be interested in the latest violated goal.
This is because the `recentGoalViolations` list contains both violations that were already fixed by previous rebalances as well as violations that still need to be addressed.
We will process only the most recent goal violation (the first element in the list when sorted by `detectionDate` in descending order) because earlier violations in the list may have already been resolved by previous rebalance operations.

> We will then convert the received JSON input into a JSON object and retrieve the `detectionDate` from the most recent entry in the `recentGoalViolations` field.

Reviewer: This sounds a strange ... received JSON, JSON input, convert JSON into JSON, etc. Maybe like this it would sound better?

```suggestion
We will parse the data from Cruise Control into a JSON object and retrieve the `detectionDate` from the most recent entry in the `recentGoalViolations` field.
```

This `detectionDate` timestamp will be used to determine whether a new auto-rebalance should be triggered.

#### Logic to trigger rebalance

To determine whether a detected anomaly should trigger a new rebalance, we need to compare the anomaly detection timestamp against the completion time of the most recent rebalance operation (whether manual or automatic).
This ensures we don't retrigger rebalances for goal violations that were already addressed.

**Tracking Rebalance Completion Times:**

> Since auto-generated KafkaRebalance resources are deleted upon completion or failure, and manual KafkaRebalance resources can be deleted by users, we need a persistent location to track completion times.

Reviewer 3: ```suggestion
Since auto-generated `KafkaRebalance` resources are deleted upon completion or failure, and manual `KafkaRebalance` resources can be deleted by users, we need a persistent location to track completion times.
```

The operator uses a new optional field in the Kafka CR status:

> **`Kafka.status.lastRebalanceCompletionTime`**

Reviewer: So, this would track auto-rebalances as well as normal balances? Or how will it work? Also, why not put it into the auto-rebalance section?

This field:
- Is updated whenever **any** KafkaRebalance resource (manual or auto-generated) transitions to a terminal state (Ready, NotReady, or Stopped)
- Persists in the Kafka CR status even after KafkaRebalance resources are deleted
- Provides a single source of truth for rebalance completion times
- Is optional (`@JsonInclude(NON_NULL)`) and only appears when at least one rebalance has completed

**Update Logic:**

When the `KafkaRebalanceAssemblyOperator` reconciles a KafkaRebalance resource and detects a transition to a terminal state:
1. Extracts the cluster name from the KafkaRebalance labels
2. Retrieves the corresponding Kafka CR
3. Updates `Kafka.status.lastRebalanceCompletionTime` to the current timestamp

> 4. Patches the Kafka CR status

Reviewer: No, this would not work and cause problems with the operators writing over each other and causing conflicts. You should follow the principle to write from one place and read from the other.

Reviewer 3: I agree with Jakub: the `KafkaAssemblyOperator` is the owner of `Kafka` CR and it should be the only one to update the `Kafka` CR status (note: the `KafkaAutoRebalanceReconciler` can do that because it's part of such operator and not an operator itself). On the other side the `KafkaRebalanceAssemblyOperator` owns `KafkaRebalance` and it can't write into `Kafka` CR. I can't see an easy solution tough. 
The `KafkaAssemblyOperator` having a watch on `KafkaRebalance` resources to detect when one ends in a terminal state to update the last rebalance time? :-/ Doing the same but poll-based on each reconciliation? :-/ (but we could lose a `KafkaRebalance` that was deleted by the user right after the rebalancing operation).

Reviewer: The obvious option is having some Config Map 🤷. I also wonder if it would make long term make sense to drop the `KafkaRebalanceAssemblyOperator and integrate the logic here. But I'm not sure I fully understand all the possible consequences. Mainly whether the fact that all rebalances would be done only if we get to the end of the reconciliation could cause any issues.

Author: I have tried to use the ConfigMap

        "unfixableViolatedGoals": [],
        "status": "FIX_STARTED"
      }
    ]
```

**Important Edge Case:** If a goal violation contains BOTH fixable goals and unfixable goals (i.e., both `fixableViolatedGoals` and `unfixableViolatedGoals` lists are non-empty), the operator will NOT trigger a rebalance. This follows Cruise Control's design principle that unfixable goals typically indicate infrastructure constraints (e.g., insufficient racks for rack-awareness, insufficient brokers for replica capacity) that cannot be resolved through partition reassignment alone. Attempting to rebalance only the fixable goals while unfixable goals remain would result in an incomplete fix and could lead to repeated rebalance attempts.

The operator can retrieve the unfixable goals from the `unfixableViolatedGoals` list in the JSON response.
When unfixable goals are detected, the operator will set a warning condition in the status of the `Kafka` CR to alert users that manual intervention is required:

```yaml
status:
  autoRebalance:
    state: Idle 
    lastTransitionTime: "2025-09-22T16:04:00Z"
    modes:
      - mode: imbalance
  conditions:
    - type: Warning
      status: "True"

>       reason: UnfixableViolatedGoal

Reviewer: This should probably have some better name to properly relate to the feature it is about.

This is **transparent to users** - no additional configuration needed.

**Mutual Exclusivity Constraint:**

Inter-broker and intra-broker rebalancing **cannot run simultaneously** (Cruise Control enforces this).
The operator handles violations sequentially:

**Priority order:**
1. Scale operations (highest)
2. Inter-broker rebalancing 
3. Intra-broker disk rebalancing (lowest)

**Execution Flow:**

When anomalies are detected, the operator categorizes violations by goal type:

1. **Both types detected**: 
   - First: Create `KafkaRebalance` with `rebalanceDisk=false` for inter-broker violations
   - After completion: Create `KafkaRebalance` with `rebalanceDisk=true` for intra-broker violations

2. **Only intra-broker violations**:
   - Create `KafkaRebalance` with `rebalanceDisk=true`
   - Goals automatically set by Cruise Control (template goals ignored)

3. **Only inter-broker violations**:
   - Standard inter-broker rebalance (existing behavior)

**State Machine Handling:**

Both inter-broker and intra-broker rebalancing use the same `RebalanceOnImbalance` state.

> The operator tracks which type of rebalance is currently running and executes them sequentially within this state:

Reviewer 3: I see a problem here. The states within an FSM are persisted into the Kafka CR so that if the cluster operator crashes it can recover from it was before. You are adding a new (internal) "state" for the `RebalanceOnImbalance` .... state :-) which makes a differentiation between `rebalanceDisk` true or false. It means that when the operator recover from a crash and it recovers the state `RebalanceOnImbalance`, it doesn't know if the `rebalanceDisk` it was handling was true or false. How are you going to solve this?

Author: I understand your concern here. You are right. I think we can either think of having two separate state RebalanceOnImbalanceInterBrokerBalancing and RebalanceOnImbalanceIntraBrokerBalancing or the other way can be introduction of new field called `pendingOperation` or `phase` to which will track the state of the `RebalanceOnImbalance` mode.

Reviewer 3: Not sure I would make the FSM more complicated by adding an additional state tbh. Maybe just storing the `rebalanceDisk` field would be enough. Let's see if others have a better idea.

* we ensure that the operator controls all rebalance and cluster remediation operations.
* using the existing `KafkaRebalance` CR system gives more visibility into what is happening and when, which (as we don't support the Cruise Control UI) enhances observability and will also aid in debugging.

### Maintenance Time Windows

Auto-rebalance on imbalance supports maintenance time windows to allow users to control when rebalancing operations occur.

> #### Implementation Design

Reviewer: Why do you need to have the maintenance windows dedicated for this? Isn't using the same maintenance window much better start?

Author: I think that way is fine too. i just thought that what if the users want to configure the maintenance window for certificates separate then auto-rebalance.

Reviewer: Apologies if I somehow made you think that we need a separate time maintenance window. I think only one is needed. Most environments do not have that many slots to fit multiple windows in. Also, the events currently handled by the maintenance window should in general be rather rare. So I would start by using the exisitng maintenance window (it was never intended as a maintenance window only for certificates - Its just that for a long time the certificates were the only type of event matching the maintenance windcow idea)

Author: Okay, I will make the changes

Reviewer 3: I totally agree, we should have just one maintenance window. Having more would be only confusing for users imho. Then it would set a precedence like in the future we'll add another window for another feature. I think if a user is doing maintenance they have to expect all the operations stopping during that time.

The reconciler currently detects any sort of scale down or scale up automatically every reconciliation.
We will configure the KafkaAutoRebalanceReconciler to detect the imbalanced cluster every reconciliation too. 
But we also need to make sure that we don't detect any goal violations when a rebalance is already ongoing (manual or automatic rebalance).
An ongoing rebalance can mean that some goal violation which might be detected in that particular reconciliation will be fixed already by the rebalance and we don't need any extra rebalance.

On every reconciliation, the operator checks the cluster state in this order and executes the **first** operation that applies:

1. **Scale-down detected** → Execute scale-down rebalance immediately (if an imbalance rebalance is running, it is stopped)
2. **Scale-up detected** → Execute scale-up rebalance immediately (if an imbalance rebalance is running, it is stopped)
3. **Goal violation detected** → Execute imbalance rebalance (only if no scale operation is in progress, and maintenance time windows are satisfied if configured)

> If a scale operation interrupts an imbalance rebalance, the imbalance rebalance is dropped.

Reviewer: Please be more clear here ... Does it mean that the operator will stop and delete an ongoing _imbalance_ rebalance to start scale-up/down rebalance? Or what does drop mean here?

Author: Yeah drop means that we will stop and delete the ongoing imbalance and start a new one. I will fix this

The operator copies over goals and rebalancing options from the referenced `template` resource to the generated rebalancing one.
If the user has not configured the anomaly detection goals in Cruise Control section of the Kafka CR then the operator will set the default goals to be used by the anomaly detector. 
The default anomaly detection goals set by the operator are `RACK_AWARENESS_GOAL`, `MIN_TOPIC_LEADERS_PER_BROKER_GOAL`, `REPLICA_CAPACITY_GOAL`, `DISK_CAPACITY_GOAL`. 

> These are similar to the default goals used for `KafkaRebalance` if the users don't mention the rebalance goals.

Reviewer: ```suggestion
These are similar to the default goals used for `KafkaRebalance` if the users don't set the rebalance goals.
```

**Tracking Rebalance Completion Times:**

Since auto-generated `KafkaRebalance` resources are deleted upon completion or failure, and manual `KafkaRebalance` resources can be deleted by users, we need a persistent location to track completion times.
The operator uses a **ConfigMap** to store this information.

**ConfigMap: `<cluster-name>-auto-rebalance-imbalance-tracker`**

This ConfigMap:
- Stores the timestamp of the most recent rebalance completion
- Tracks the current rebalancing phase for JBOD clusters (inter-broker vs. intra-broker)
- Is created in the same namespace as the Kafka cluster
- Persists even after `KafkaRebalance` resources are deleted
- Allows the operator to recover state after crashes or restarts

> - Is managed exclusively by the `KafkaAssemblyOperator` (maintains proper ownership boundaries)

Reviewer: Really? Wouldn't it be `KafkaRebalanceAssemblyOperator` if it writes all rebalances?

**ConfigMap Lifecycle:**

- **Creation**: Created automatically by the operator when the first rebalance (manual or auto-generated) completes

> - **Updates**: Updated throughout the lifecycle whenever rebalances complete or JBOD rebalancing phases change

Reviewer: What does `JBOD rebalancing phases change` mean? Please elaborate.

Author: By that I meant the interbroker and intrabroker rebalance phases that can happen when both type of goals are violated and detected in JBOD supporting cluster

Reviewer: Please make sure you don't answer it just here but that you make the things clear in the proposal.

- **Persistence**: The ConfigMap is **not deleted** when rebalances complete or when the Kafka cluster is idle
- **Deletion**: Only deleted when the Kafka cluster itself is deleted (via owner reference to the Kafka CR)
- **Rationale**: Persisting the ConfigMap allows accurate timestamp comparison across reconciliations and operator restarts, preventing duplicate rebalancing operations

Example ConfigMap:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-cluster-auto-rebalance-imbalance-tracker
  namespace: kafka
  labels:
    strimzi.io/cluster: my-cluster
    strimzi.io/kind: Kafka
data:
  lastRebalanceCompletionTime: "2026-06-03T14:23:45Z"
  # For JBOD clusters in RebalanceOnImbalance state, tracks which phase is active
  rebalancingPhase: "inter-broker"  # Values: "inter-broker", "intra-broker", or absent
```

**ConfigMap Data Fields:**

1. **`lastRebalanceCompletionTime`**: ISO-8601 timestamp of the most recent rebalance completion (manual or auto-generated)

> 2. **`rebalancingPhase`** (optional, only present during active RebalanceOnImbalance for JBOD clusters):

Reviewer: Above you are talking about how the operator will list KafkaRebalances. But is it really needed if the ConfigMap tracks if something is in progress?

   - `"inter-broker"`: Inter-broker rebalancing is currently active
   - `"intra-broker"`: Intra-broker disk rebalancing is currently active
   - Field is absent when not in RebalanceOnImbalance state or for non-JBOD clusters

**Update Logic:**

To maintain proper operator ownership boundaries (where `KafkaAssemblyOperator` owns the `Kafka` CR and `KafkaRebalanceAssemblyOperator` owns `KafkaRebalance` CRs), the update mechanism works as follows:

> The `KafkaAssemblyOperator` (which contains `KafkaAutoRebalanceReconciler`) maintains a watch on `KafkaRebalance` resources with the `strimzi.io/cluster` label matching Kafka clusters it manages.

Reviewer: Are you sure? KafkaAssemblyOperator does not have such a watch. KafkaRebalanceAssemblyOperator does.

Author: Yes, I was thinking of having a watch in the Kafka Assembly Operator just like we have in the KafkaRebalanceAssemblyOperator but you are correct we don't need it. We can just make use of the KafkaRebalanceAssemblyOperator. Sorry I think I just tried to do everything with KafkaAssemblyOperator and the KafkaAutoRebalance Reconciler.

When any `KafkaRebalance` resource (manual or auto-generated) transitions to a terminal state (Ready, NotReady, or Stopped):
1. The watch trigger fires in `KafkaAssemblyOperator`
2. The operator extracts the cluster name from the `KafkaRebalance` labels (`strimzi.io/cluster`)
3. The operator updates the ConfigMap `<cluster-name>-auto-rebalance-imbalance-tracker` with the current timestamp in the `lastRebalanceCompletionTime` field
4. This happens **before** the `KafkaRebalance` resource is deleted (for auto-generated ones)

This approach:
- Preserves operator ownership boundaries: only `KafkaAssemblyOperator` modifies the state ConfigMap
- Captures completion times for both manual and auto-generated rebalances
- Works even if users delete `KafkaRebalance` resources manually (the watch sees the terminal state before deletion)
- Provides a separate, queryable resource for tracking rebalance state without polluting the Kafka CR status

**Comparison Logic:**

A rebalance on imbalance will be triggered when:
```java

> ConfigMap stateConfigMap = getConfigMap(namespace, clusterName + "-auto-rebalance-imbalance-tracker");

Reviewer: Doesn't it track all rebalances? The name should correspond to it.

Author: Does this sound good `cluster-name-rebalance-on-imbalance-tracker` ?

Reviewer: But it does not track imbalances, or? It tracks all rebalances if I read it right.

Reviewer 3: Yeah I added a similar comment in a previous section when the ConfigMap is introduced.

When an auto-rebalance on imbalance fails (the `KafkaRebalance` resource transitions to `NotReady` state):

1. **ConfigMap updated automatically**: The `KafkaRebalanceAssemblyOperator` updates the `<cluster-name>-rebalancing-tracker` ConfigMap's `lastRebalanceCompletionTime` when it reconciles the `NotReady` state

> 2. **Error propagation**: The `KafkaAutoRebalanceReconciler` sets a Warning condition on the Kafka CR with the error message from the `KafkaRebalance` status

Reviewer: I think I raised it before. But this has to be logged in the first place as the condition will probably disappear right away? Also, if you want to add condition, you should exlain what condition would that be. (e.g. by sharing an example)

Author: I think logging seems fine but I think it would still make sense to generate a condition. If the anomaly is not fixed then the warning will keep generating. I think even if somehow the user misses the logs, they will eventually catch the warning in the CR as it will be there till the reconcilation. I will add an example

Reviewer: It is not like it is a blocker. But unless the anomaly is persistent in Cruise Control (which as far as I understood it is not and it will only eventually show up again), the warning condition will be there for few seconds only.

* You update the resource by adding the condition
* That triggers a new reconciliation right away
* The new reconciliation with nothing else to do otherwise, will right away wipe the condition
* Which triggers another reconciliation within few seconds which this time will not do anything anymore and just burn CPU

Reviewer 3: If on next reconciliation, when the operator queries the CC anomaly detector state, the anomaly is not there anymore then Jakub is right. The reconciliation process will just wipe out the condition (or at least this is my expectation) because there are no anomalies. I would expect the anomaly to be there as result of the query but maybe you are just discarding it because of it's before the last timestamp stored in the ConfigMap?

Author: The anomalies are persistent in Cruise Control unless they are resolved. As @ppatierno said, we only consider those anomaly which have  the last timestamp greater than the timestamp stored in the ConfigMap which means a fresh anomaly. So we will reject the anomalies caught earlier. We will detect it again in the next cycle though. I think it makes sense to just log it then

### Maintenance Time Windows

> Auto-rebalance on imbalance can respect maintenance time windows to control when rebalancing operations occur, using the existing `Kafka.spec.maintenanceTimeWindows` configuration that is already used for certificate renewals in Strimzi.

Reviewer: What does it mean that it `can` respect? It either respects it or it does not respect it.

Author: Fixing this

   ```yaml
   spec:
     # maintenanceTimeWindows not specified
     cruiseControl:
       autoRebalance:
         - mode: imbalance
   ```
   - Rebalancing triggers immediately when anomalies are detected
   - No time-based restrictions
   - Suitable for development/test clusters or critical production clusters requiring immediate action

**Behavior:**

- **When maintenance windows are configured**:
  - When an anomaly is detected outside a maintenance window, the operator records the detection but does NOT trigger a rebalance
  - During the next reconciliation that falls within a maintenance window, the operator queries Cruise Control's `state` endpoint
  - If the anomaly still exists (based on `detectionDate` comparison against `lastRebalanceCompletionTime`), the rebalance is triggered
  - The anomaly detection continues every reconciliation, but rebalance triggering is gated by the maintenance window check using `Util.isMaintenanceTimeWindowsSatisfied(reconciliation, maintenanceWindows, clock.instant())`

- **When no maintenance windows are configured**:
  - Rebalance is triggered immediately when a fixable goal violation is detected

>   - Useful for critical clusters where imbalance correction cannot wait

Reviewer: I would remove this. I think it is exactly the critical clusters where you cannot afford to run the rebalances anytime.

**Emergency Override:**

If users need immediate rebalancing despite configured maintenance windows (e.g., critical disk capacity issue), they can:

> 1. Temporarily remove the `maintenanceTimeWindows` field from the `Kafka` CR, or

Reviewer: Should we keep just the second option? While the users can surely remove the window temporarily, it sounds like a fragile thing where one day someone forgets to add it back.

Author: Sure

Reviewer 3: +1

2. Create a manual `KafkaRebalance` resource (which bypasses maintenance windows)

#### Interaction with Scale Operations

Scale-up and scale-down auto-rebalance operations will continue to **ignore** maintenance windows (executed immediately), while imbalance auto-rebalance respects them.

> This design follows the principle that scaling operations are typically urgent infrastructure changes (adding capacity for increased load or removing failed/excess nodes), whereas imbalance correction is a performance optimization that can be deferred to low-traffic periods defined by maintenance windows.

Reviewer: I do not have a major problem with the scaling operations having higher priority. But I do not think this explanation makes sense and justifies it in any way. What urgent issues that the scaling would be solving are you talking abotu here? If there are any, please add exact examples.

Author: I can remove this explanation if it doesn't make sense. I dont have any exact examples

Reviewer: I would maybe remove it then. Because at least I personally do not see scaling as a solution to urgent changes and have never seen it (while for stateless apps you might see scaling for demand as _urgent_ thing, in Kafka it would actually slow you down in the first place, so I do not think it would be really an urgent way).

Just to be clear, I'm fine with the logic to do scaling first, I think it makes roughly some sense. Just the explanation seems the exact opposite to how things work form my experience.

Reviewer 3: Yeah I agree as well. Having scaling operations with higher priority is fine but the I can't see the explanation above so much valid in real use cases.

- Inter-broker: Uses goals from template
- Intra-broker: Ignores goals from template (auto-set), but uses `concurrentIntraBrokerPartitionMovements` if specified

> **Operator Crash Recovery:**

Reviewer: I'm a bit confused. What crash are we talking about here? What do you expect to loose that you need to recover? You should have the Kafka CR state and the KafkaRebalance resource to recover any state. And in many other places we seem to just throw away the issues detected by Cruise Control with the idea that if thea are still a problem they will be raised again. So why do we need some complicated recovery?

Author: I was trying to address this problem -> https://github.com/strimzi/proposals/pull/211#discussion_r3297717114

Author: @scholzj Do the shared comment resolves your question. I think we want it because the FSM is persisted into KafkaCR which means that when the operator crashes and recovers it will start from `RebalanceOnImbalance` which can be correspond to both Intrabroker balancing and InInterBrokerRebalancing. So how will the operator know which rebalance to proceed with?

Author: @ppatierno a quick question on the same, in case the operator just proceeded with Interbroker Rebalance and lets say skipped IntrabrokerRebalance.  CC will again detect the intrabroker rebalance again and in the next cycle and fix it. Wdyt can be the problem in case we don't have any recovery mechanism?

Reviewer 3: That could work but it's something you should try to test somehow. I can imagine you run a useless rebalancing (inter broker) while the CC was doing an intra-broker movements (but we lost the status of it). Or is it something you can get from a CC endpoint to understand which one it was doing?

Reviewer: @ShubhamRwt No, not really. It did not explained to me why it is needed. That comment only makes me wonder why don't you have two states, e.g.: `IntraRebalanceOnImbalance` and `InterRebalanceOnImbalance`

Author: Hi @scholzj, Sorry for not relying early. Actually adding a new state would cause more complexity to the FSM and I was try to avoid that. I am for now checking what if in case of a crash, we can just do a inter broker rebalance instead of the checking what rebalance is left. CC will then pickup the intra broker violation again in the next cycle. I was trying to test it with my POC but it is taking some time since I need to make some code changes

Reviewer: Ok, if the additional state adds too much complexity, lets recover from the Config Map.

The second part is something I'm more curious about. to me, it seems tht there are many places where we treat it as _If we do not act on the Cruise Control finding right away, it will be raised up agian and we do it next time_. And then suddenly here we have a complicated recovery process. My expectation would have been that we just check the state and if it is not idle, we see what KafkaRebalance resource we have out there and what state it is in to continue with it.

Author: so lets say 

We have a JBOD clusters with both inter-broker AND intra-broker violations:

  1. Create KafkaRebalance with rebalanceDisk=false (inter-broker)
  2. Wait for KafkaRebalance to move to Ready
  3. Delete the KafkaRebalance
  4. Create new KafkaRebalance with rebalanceDisk=true (intra-broker)
  5. Wait for completion
  6. Transition to Idle
  
Lets say in case, the operator crashes after step 3

Now we already have done the inter-broker rebalance and the rebalance resource is deleted. In case the operator crashed, we would still stay in `RebalanceOnImbalance` state but we don't know what rebalance to proceed with next. 

As we said in the proposal, if both violation are detected, we will stay in `RebalanceOnImbalance` state unless both inter-broker rebalance and intra-broker rebalance has executed where the inter-broker rebalance runs first and then  intra-broker rebalance.

Reviewer: If there is still a violation after step 3, why would it not be raised again later by CC? Why do we here need to remember it and recover from it when with the regular rebalance or scale up/down we just say we ignore it and if there is still a problem it will be raised again but not here?

Author: It would be raised again later by CC for sure. I think the problem was which rebalance would be performed after we recover. After the recovery we will be on `RebalanceonImbalance` state. Since no KafkaRebalance was generated as the operator crashed after 3rd step. We don't know what rebalance to perform or what state we are in. 

I think we need to decide what we want to do in that case:

1. If operator crashes during RebalanceOnImbalance and the kafkaRebalance is null(i.e. we dont know what rebalance happed or not), we can either just skip everything and move to Idle. This is kind of in line with how we say that CC can detect the issue again in next cycle. 

2. The other option is to track the phases and that way we know after recovery what rebalance is already done

3. Third option is that after recovery we can create the kafka rebalance again in interbroker rebalance mode i.e we start from the beginning. 
   - CC will not be doing doing any new rebalance in case the goals are already satisfied
  
 4. Have 2 states but that would make the FSM complex:
   - more states means more operation to perform
      Think if we have scale up and scale down existing and then we also get RebalanceonIntrabrokerImbalance and RebalanceonInterbrokerImbalance.

Reviewer: To me it seems better to just move to idle if you have the state but the KafkaRebalance was deleted already / not created yet. If CC throws the issue again next time, it lowers the complexity of everything.

Author: Yes, I think that will be the best way. If  KafkaRebalance was deleted already / not created yet then we just mov the state to Idle else we continue with the rebalance

Author: @ppatierno wdyt?

Reviewer 3: I think it's reasonable, yes.

The following changes need to be made to support this feature:

> 1. **ConfigMap for state tracking**: Create a new ConfigMap `<cluster-name>-rebalancing-tracker` to track the completion time of the most recent rebalance operation (manual or automatic). This ConfigMap contains a single data entry: `lastRebalanceCompletionTime`.

Reviewer: Not an API change really. API are the custom resources. Not an internal config map.

Author: Fixing

2. **Add `imbalance` mode**: Support automatic rebalancing on goal violations (handles both inter-broker and intra-broker for JBOD) in the `Kafka.spec.cruiseControl.autoRebalance` configuration.

3. **Add `RebalanceOnImbalance` state**: New FSM state when rebalancing (inter-broker or intra-broker) triggered by goal violations is running.

> 4. **Add intra-broker disk goals**: For JBOD storage support: `INTRA_BROKER_DISK_CAPACITY_GOAL` and `INTRA_BROKER_DISK_USAGE_DISTRIBUTION_GOAL`.

Reviewer: Where exactly int he API would this be added? Not clear.

Author: I dont think it exactly an API change. I will fix this section

> 5. **Reuse existing maintenance windows**: The feature will leverage the existing `Kafka.spec.maintenanceTimeWindows` field and `Util.isMaintenanceTimeWindowsSatisfied()` method already used for certificate renewals.

Reviewer: I guess this is not an API change:

Author: Fixing this
